# Phyloscanner analysis of deep-sequence trees from a population-based sample

### Introduction: batches and input data
It is computationally very challenging to reconstruct viral trees from all
deep-sequence reads of hundreds or thousands of individuals. To
address this challenge, we divide a large population-based sample into batches
of 50 to 75 individuals, and then run *phyloscanner* on all possible
pairs of batches to generate read alignments and deep-sequence phylogenies. 

This tutorial describes the steps to infer phylogenetic deep-sequence
relationships from a large number of deep-sequence phylogenies of batches of
individuals. 

The code below assumes that *phyloscanner_make_trees.py* was already run to
generate read alignments and deep-sequence phylogenies for each batch, and the
following file structure.
1. Each batch is identified with the prefix `ptyrX_` where `X` is an integer.
2. Three files are available for each batch. First, a text file listing all
   individuals in the batch, in file `ptyrX_patients.txt`. 
3. Second, the read alignments generated with *phyloscanner*, zipped into file
   `ptyrX_trees_fasta.zip`. 
4. Third, the deep-sequence phylogenies generated with *phyloscanner*, zipped
   into file `ptyrX_trees_newick.zip`. 

Data Set S1 contains these files for 345 batches of individuals of the Rakai
population-based sample.

### Setting up your work environment

Start by defining the base directories for your project: 
1. `HOME` Base directory.
2. `in.dir` Name of directory containing read alignments and deep-sequence
   trees, which can be generated from *bam* files with *phyloscanner*. If you
   would like to start with Data Set S1, you should unzip files in Data Set S1
   to this directory.
3. `out.dir` Name of output directory.
4. `work.dir` Name of temp directory.
5. `prog.pty` Full path to the *phyloscanner* program
   *phyloscanner_make_trees.py*
    
For example:    
```{r, eval=FALSE, message=FALSE, warning= FALSE, include=TRUE} 
require(Phyloscanner.R.utilities)

HOME		<- '/Users/Oliver/Dropbox (SPH Imperial College)/2015_PANGEA_DualPairsFromFastQIVA'								
in.dir		<- file.path(HOME,'RakaiPopSample_deepseqtrees')	
out.dir		<- file.path(HOME,"RakaiPopSample_phyloscanner_out")
work.dir	<- file.path(HOME,"RakaiPopSample_phyloscanner_work")				
prog.pty	<- '/Users/Oliver/git/phylotypes/phyloscanner_make_trees.py'

# create directories if they dont exist
dir.create(out.dir, showWarnings=FALSE)
dir.create(work.dir, showWarnings=FALSE)
```  

### Prepare bash scripts to run phyloscanner

The next step is to define the input arguments to *phyloscanner*. Please see the
*phyloscanner* manual for details. The default arguments that were used for
analysis of the Rakai population-based sample are as follows. 
```{r, eval=FALSE, message=FALSE, warning=FALSE, include=TRUE} 
pty.args			<- list(	prog.pty=prog.pty, 
		prog.mafft=NA, 
		prog.raxml=NA, 
		data.dir=NA, 
		work.dir=work.dir, 
		out.dir=out.dir, 
		alignments.file=system.file(package="Phyloscanner.R.utilities", "HIV1_compendium_AD_B_CPX_v2.fasta"),
		alignments.root='REF_CPX_AF460972', 
		alignments.pairwise.to='REF_B_K03455',
		bl.normalising.reference.file=system.file(package="Phyloscanner.R.utilities", "data", "hiv.hxb2.norm.constants.rda"),
		bl.normalising.reference.var='MEDIAN_PWD',														
		window.automatic= '', 
		merge.threshold=0, 
		min.read.count=1, 
		quality.trim.ends=23, 
		min.internal.quality=23, 
		merge.paired.reads=TRUE, 
		no.trees=FALSE, 
		dont.check.duplicates=FALSE,
		dont.check.recombination=TRUE,
		num.bootstraps=1,
		all.bootstrap.trees=TRUE,
		strip.max.len=350, 
		min.ureads.individual=NA, 
		win=c(800,9400,25,250), 				
		keep.overhangs=FALSE,
		use.blacklisters=c('ParsimonyBasedBlacklister','DownsampleReads'),
		tip.regex='^(.*)_fq[0-9]+_read_([0-9]+)_count_([0-9]+)$',
		roguesubtree.kParam=20,
		roguesubtree.prop.threshold=0,
		roguesubtree.read.threshold=20,
		dwns.maxReadsPerPatient=50,	
		multifurcation.threshold=1e-5,
		split.rule='s',
		split.kParam=20,
		split.proximityThreshold=0,
		split.readCountsMatterOnZeroBranches=TRUE,
		split.pruneBlacklist=FALSE,
		trms.allowMultiTrans=TRUE,
		pw.trmw.min.reads=30,									
		pw.trmw.min.tips=1,
		pw.trmw.close.brl=0.025,
		pw.trmw.distant.brl=0.05,
		pw.prior.keff=2,
		pw.prior.neff=3,
		pw.prior.keff.dir=2,
		pw.prior.neff.dir=3,				
		pw.prior.calibrated.prob=0.66,
		mem.save=0,
		verbose=TRUE,				
		select=NA 
)	
save(pty.args, file=file.path(out.dir, 'pty.args.rda'))
```
  
Next, we will prepare UNIX *bash* scripts to run a large number of
*phyloscanner* analyses on a population-based sample. Each *bash* script
corresponds to the deep-sequence phylogenetic analysis of one batch of
individuals that are analysed jointly.

For each batch of individuals to be processed, find the corresponding list of
patients in the input directory: 
```{r, eval=FALSE, message=FALSE, warning=FALSE, include=TRUE}
pty.c	<- data.table(FILE_BAM=list.files(in.dir, pattern='_patients.txt', full.names=TRUE))
pty.c[, PTY_RUN:= as.integer(gsub('ptyr','',gsub('_patients.txt','',basename(FILE_BAM))))]
```

Check which (if any) batches have already been processed, and remove them from
the TODO list:
```{r, eval=FALSE, message=FALSE, warning=FALSE, include=TRUE}
tmp		<- data.table(FILE_TRMW=list.files(out.dir, pattern='_pairwise_relationships.rda', full.names=TRUE))
tmp[, PTY_RUN:= as.integer(gsub('ptyr','',gsub('_pairwise_relationships.rda','',basename(FILE_TRMW))))]
pty.c	<- merge(pty.c, tmp, by='PTY_RUN', all.x=1)
pty.c	<- subset(pty.c, is.na(FILE_TRMW))
```

For each batch of individuals, create a UNIX *bash* script to run
*phyloscanner*: 
```{r, eval=FALSE, message=FALSE, warning=FALSE, include=TRUE}
setkey(pty.c, PTY_RUN)		
pty.c	<- pty.c[, { 
			prefix.infiles	<- gsub('patients.txt','',FILE_BAM)			
			cmd				<- phsc.cmd.phyloscanner.one.resume(prefix.infiles, pty.args)
			list(CMD=cmd) 
		}, by='PTY_RUN']		
pty.c[1,cat(CMD)]		
```

Each bash script should look similar to this:
```{r, eval=FALSE, message=FALSE, warning=FALSE, include=TRUE}
'
CWD=$(pwd)
echo $CWD
mkdir -p "$CWD/pty_18-10-17-10-03-47"
cp "/Users/Oliver/Dropbox (SPH Imperial College)/2015_PANGEA_DualPairsFromFastQIVA/RakaiPopSample_deepseqtrees/ptyr223_patients.txt" "$CWD/pty_18-10-17-10-03-47"
unzip "/Users/Oliver/Dropbox (SPH Imperial College)/2015_PANGEA_DualPairsFromFastQIVA/RakaiPopSample_deepseqtrees/ptyr223_trees_fasta.zip" -d "$CWD/pty_18-10-17-10-03-47"
unzip "/Users/Oliver/Dropbox (SPH Imperial College)/2015_PANGEA_DualPairsFromFastQIVA/RakaiPopSample_deepseqtrees/ptyr223_trees_newick.zip" -d "$CWD/pty_18-10-17-10-03-47"
cd "$CWD/pty_18-10-17-10-03-47"
Rscript /Users/Oliver/git/phylotypes/deprecated/NormalisationLookupWriter.R --scriptdir /Users/Oliver/git/phylotypes/deprecated "$CWD/pty_18-10-17-10-03-47/ptyr223_InWindow_" "/Users/Oliver/Library/R/3.3/library/Phyloscanner.R.utilities/data/hiv.hxb2.norm.constants.rda" "$CWD/pty_18-10-17-10-03-47/ptyr223_normconst.csv" "MEDIAN_PWD"  --standardize
Rscript /Users/Oliver/git/phylotypes/tools/parsimony_based_blacklister.R 20 0 20 "$CWD/pty_18-10-17-10-03-47/ptyr223_InWindow_" "$CWD/pty_18-10-17-10-03-47/ptyr223_blacklistsank_InWindow" --dualsOutputFile "$CWD/pty_18-10-17-10-03-47/ptyr223_duallistsank_InWindow" --outgroupName REF_CPX_AF460972 --tipRegex "^(.*)_fq[0-9]+_read_([0-9]+)_count_([0-9]+)$" --multifurcationThreshold 1e-05  --branchLengthNormalisation "$CWD/pty_18-10-17-10-03-47/ptyr223_normconst.csv" --verbose
Rscript /Users/Oliver/git/phylotypes/tools/downsample_reads.R 50 $CWD/pty_18-10-17-10-03-47/ptyr223_ $CWD/pty_18-10-17-10-03-47/ptyr223_blacklistdwns_ --blacklist $CWD/pty_18-10-17-10-03-47/ptyr223_blacklistsank_InWindow_ --tipRegex "^(.*)_fq[0-9]+_read_([0-9]+)_count_([0-9]+)$" --seed 42 --verbose
Rscript /Users/Oliver/git/phylotypes/tools/split_hosts_to_subgraphs.R "$CWD/pty_18-10-17-10-03-47/ptyr223_" "ptyr223" --blacklist "$CWD/pty_18-10-17-10-03-47/ptyr223_blacklistdwns_" --outputdir "$CWD/pty_18-10-17-10-03-47" --idFile "$CWD/pty_18-10-17-10-03-47/ptyr223_patients.txt" --outgroupName REF_CPX_AF460972 --splitsRule s --kParam 20 --proximityThreshold 0 --readCountsMatterOnZeroBranches --tipRegex "^(.*)_fq[0-9]+_read_([0-9]+)_count_([0-9]+)$" --multifurcationThreshold 1e-05 --branchLengthNormalisation "$CWD/pty_18-10-17-10-03-47/ptyr223_normconst.csv" --outputAsRDA --pdfwidth 30 --pdfrelheight 0.15 --verbose
Rscript /Users/Oliver/git/phylotypes/tools/classify_relationships.R "$CWD/pty_18-10-17-10-03-47/ProcessedTree_s_ptyr223_" "$CWD/pty_18-10-17-10-03-47/subgraphs_s_ptyr223_" "$CWD/pty_18-10-17-10-03-47/ptyr223" --branchLengthNormalisation "$CWD/pty_18-10-17-10-03-47/ptyr223_normconst.csv" --verbose
Rscript /Users/Oliver/git/phylotypes/tools/summary_statistics.R --scriptDir /Users/Oliver/git/phylotypes/tools "$CWD/pty_18-10-17-10-03-47/ptyr223_patients.txt" "$CWD/pty_18-10-17-10-03-47/ProcessedTree_s_ptyr223_InWindow_" "$CWD/pty_18-10-17-10-03-47/subgraphs_s_ptyr223_InWindow_" "$CWD/pty_18-10-17-10-03-47/ptyr223_" --tipRegex "^(.*)_fq[0-9]+_read_([0-9]+)_count_([0-9]+)$" --blacklists "$CWD/pty_18-10-17-10-03-47/ptyr223_blacklistdwns_InWindow_" --verbose
Rscript /Users/Oliver/git/phylotypes/deprecated/TransmissionSummary.R "$CWD/pty_18-10-17-10-03-47/ptyr223_patients.txt" "$CWD/pty_18-10-17-10-03-47/ptyr223_classification_InWindow_" "$CWD/pty_18-10-17-10-03-47/ptyr223_trmStats.csv" --scriptdir /Users/Oliver/git/phylotypes/deprecated --summaryFile "$CWD/pty_18-10-17-10-03-47/ptyr223_patStatsFull.csv" --minThreshold 1 --detailedOutput "$CWD/pty_18-10-17-10-03-47/ptyr223_trmStatsPerWindow.rda" --allowMultiTrans --verbose
Rscript /Users/Oliver/Library/R/3.3/library/Phyloscanner.R.utilities/phsc.pairwise.relationships.Rscript --infile "$CWD/pty_18-10-17-10-03-47/ptyr223_trmStatsPerWindow.rda" --outfile "$CWD/pty_18-10-17-10-03-47/ptyr223_pairwise_relationships.rda" --trmw.min.reads 30 --trmw.min.tips 1 --trmw.close.brl 0.025 --trmw.distant.brl 0.05 --prior.keff 2 --prior.neff 3 --prior.keff.dir 2 --prior.neff.dir 3 --prior.calibrated.prob 0.66 --rel.likely.pair --rel.likely.pair.by.distance.only --rel.likely.pair.by.topology.only --rel.likely.pair.by.cross.table --rel.direction --rel.chain
Rscript /Users/Oliver/Library/R/3.3/library/Phyloscanner.R.utilities/phsc.read.processed.phyloscanner.output.in.directory.Rscript --prefix.infiles "$CWD/pty_18-10-17-10-03-47/ptyr223_" --save.file.base "$CWD/pty_18-10-17-10-03-47/ptyr223_" --read.likelytransmissions --read.trees --read.subtrees --zip
mv ptyr223* "/Users/Oliver/Dropbox (SPH Imperial College)/2015_PANGEA_DualPairsFromFastQIVA/RakaiPopSample_phyloscanner_out"
for file in *; do
   zip -ur9XTj ptyr223_otherstuff.zip "$file"
done
mv ptyr223_otherstuff.zip "/Users/Oliver/Dropbox (SPH Imperial College)/2015_PANGEA_DualPairsFromFastQIVA/RakaiPopSample_phyloscanner_out"
cd $CWD
rm -r "$CWD/pty_18-10-17-10-03-47"
NULL
'		
```
	

