<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> • Phyloscanner.R.utilities</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Phyloscanner.R.utilities</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Rakai.01.data_description.html">Data description</a>
    </li>
    <li>
      <a href="articles/Rakai.02.run_phyloscanner.html">Running *phyloscanner* in high-throughput on a population-based sample</a>
    </li>
    <li>
      <a href="articles/Rakai.03.reconstruct_transmission_networks.html">Reconstructing transmission networks</a>
    </li>
    <li>
      <a href="articles/Rakai.04.direction_of_transmission.html">Inferring the direction of transmission</a>
    </li>
    <li>
      <a href="articles/Stage1.create.read.alignments.html">Stage 1 analysis: identify phylogenetically closely related individuals</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/olli0601/Phyloscanner.R.utilities">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
    
<div id="phyloscanner-r-utilities" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#phyloscanner-r-utilities" class="anchor"></a>Phyloscanner.R.utilities</h1></div>
</div>
<div id="utility-functions-for-deep-sequence-phylogenetic-analyses-with-phyloscanner" class="section level1">
<h1 class="hasAnchor">
<a href="#utility-functions-for-deep-sequence-phylogenetic-analyses-with-phyloscanner" class="anchor"></a>Utility functions for deep-sequence phylogenetic analyses with <em>phyloscanner</em>
</h1>
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p><em>Phyloscanner.R.utilities</em> is a set of functions used to help apply <a href="https://github.com/BDI-pathogens/phyloscanner">the phyloscanner software</a> to large datasets of a population-based sample, and reconstruct transmission networks from multiple <em>phyloscanner</em> runs. Please go to the <a href="https://github.com/BDI-pathogens/phyloscanner"><em>phyloscanner</em> github page</a> for access to the main software package.</p>
<p>The software package comprises</p>
<ol>
<li>R functions for setting up and running a large number of <em>phyloscanner</em> analyses on batches of deep-sequence data in parallel.</li>
<li>R functions for collecting <em>phyloscanner</em> output of multiple batches, and for reconstructing transmission networks.</li>
<li>R functions to visualise reconstructed transmission networks and primary <em>phyloscanner</em> output.</li>
</ol>
</div>
<div id="system-requirements" class="section level2">
<h2 class="hasAnchor">
<a href="#system-requirements" class="anchor"></a>System requirements</h2>
<p>Both <em>phyloscanner</em> and <em>Phyloscanner.R.utilities</em> are supported on <em>Linux</em> and <em>MacOS</em>.</p>
<ol>
<li>
<em>phyloscanner</em> builds on standard tools for deep-sequence data analysis and phylogeny reconstruction, such as <em>samtools</em> and <em>RAxML</em>. Please install these first; <a href="https://github.com/BDI-pathogens/phyloscanner/blob/master/InfoAndInputs/InstallationNotesForMakingTrees.sh">instructions are available here for <em>Linux</em> and <em>MacOS</em></a>. If you have installation issues, <a href="https://github.com/BDI-pathogens/phyloscanner/issues">please report it here and we will get back to you</a>. This step may take up to 60 minutes.</li>
<li>
<p><em>phyloscanner</em> and <em>Phyloscanner.R.utilities</em> depend on several <em>R</em> packages. We find it easiest to install them as follows. First run the following <code>install_github</code> command, and then install any packages that could not be installed manually (these packages not on CRAN and need to be installed from Bioconductor/ github) That is:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools:::<span class="kw">install_github</span>(<span class="st">"olli0601/Phyloscanner.R.utilities"</span>, <span class="dt">dependencies=</span><span class="ot">TRUE</span>, <span class="dt">build_vignettes=</span><span class="ot">FALSE</span>)</code></pre></div>
<p>On a fresh R build, installation will fail with</p>
<pre class="text"><code>ERROR: dependencies ‘ggtree’, ‘ggnet’, ‘Rsamtools’, ‘RBGL’ are not available</code></pre>
<p>You will then need</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">source</span>(<span class="st">"https://bioconductor.org/biocLite.R"</span>)
<span class="kw">biocLite</span>(<span class="kw">c</span>(<span class="st">"Rsamtools"</span>,<span class="st">"RBGL"</span>,<span class="st">"ggtree"</span>),<span class="dt">dependencies=</span><span class="ot">TRUE</span>, <span class="dt">build_vignettes=</span><span class="ot">FALSE</span>)
devtools:::<span class="kw">install_github</span>(<span class="st">"briatte/ggnet"</span>, <span class="dt">dependencies=</span><span class="ot">TRUE</span>)</code></pre></div>
<p>This step may take up to 60 minutes. If you have issues with installation/running of <em>Phyloscanner.R.utilities</em>, <a href="https://github.com/olli0601/Phyloscanner.R.utilities/issues">please report it here and we will get back to you</a>.</p>
</li>
</ol>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<ol>
<li>
<a href="https://github.com/olli0601/Phyloscanner.R.utilities/tree/master/misc/phyloscanner_v1.1.2.tar.gz">Download <em>phyloscanner</em> version 1.1.2</a> and unzip to a directory of your choice.</li>
<li>
<p>Install <em>Phyloscanner.R.utilities</em> in R:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools:::<span class="kw">install_github</span>(<span class="st">"olli0601/Phyloscanner.R.utilities"</span>, <span class="dt">dependencies=</span><span class="ot">TRUE</span>, <span class="dt">build_vignettes=</span><span class="ot">FALSE</span>)
<span class="kw">require</span>(Phyloscanner.R.utilities)</code></pre></div>
<p>This step takes less than 5 minutes. If you have issues with installation/running of <em>Phyloscanner.R.utilities</em>, <a href="https://github.com/olli0601/Phyloscanner.R.utilities/issues">please report it here and we will get back to you</a>.</p>
</li>
</ol>
</div>
<div id="general-protocol-for-analyses-at-the-population-level" class="section level2">
<h2 class="hasAnchor">
<a href="#general-protocol-for-analyses-at-the-population-level" class="anchor"></a>General protocol for analyses at the population-level</h2>
<p>It is computationally challenging to reconstruct viral trees from deep-sequence reads of hundreds or thousands of individuals. To address this challenge, we generally proceed in two stages.</p>
<p><strong>In a first stage</strong>, the large population-based sample is divided into groups of 50 to 75 individuals, and then <em>phyloscanner</em> is run on all possible pairs of groups to generate read alignments and deep-sequence phylogenies. From these trees, potentially phylogenetically close pairs are identifed and from those, networks of individuals that were connected through at least one common, phylogenetically close individual.</p>
<p><strong>In a second stage</strong>, <em>phyloscanner</em> analyses are repeated on all individuals that together form a potential transmission network. In addition to reads from all individuals in a potential transmission network, each analysis also includes reads from closely related individuals that act as control sequences. This step allows us to resolve the ordering of transmission events within transmission networks, and to confirm potential transmission pairs within a network.</p>
<p><strong>Both stages 1 and 2 consist of very similar steps.</strong> In either stage, phylogenetic trees are first generated from deep-sequence reads of batches of individuals (either from two randomly paired groups of individuals in stage 1, or phylogenetically closely connected individuals in a potential transmission network in stage 2). Thereafter, phylogenetic relationships between these individuals are estimated from the reconstructed deep-sequence trees.</p>
<p><strong>The similarity of the two stages allows us to use the same high-throughput R scripts for analysis</strong></p>
</div>
<div id="tutorials-for-phyloscanner-analyses-of-large-population-based-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#tutorials-for-phyloscanner-analyses-of-large-population-based-samples" class="anchor"></a>Tutorials for <em>phyloscanner</em> analyses of large population-based samples</h2>
<p>To demonstrate analysis of a deep-sequence reads from of large population-based sample of individuals, we here provide several tutorials on data from Rakai District, Uganda. <a href="articles/Rakai.01.data_description.html">The data are described here.</a></p>
<div id="running-phyloscanner-in-high-throughput" class="section level3">
<h3 class="hasAnchor">
<a href="#running-phyloscanner-in-high-throughput" class="anchor"></a><a href="articles/Rakai.02.run_phyloscanner.html">Running <em>phyloscanner</em> in high-throughput</a>
</h3>
<p><a href="articles/Rakai.02.run_phyloscanner.html">This tutorial</a> starts with <em>phyloscanner</em> analyses of deep-sequence phylogenetic trees for each potential transmission network identified on this population-based sample of 2,652 infected individuals. The main objective is to illustrate how large numbers of phyloscanner runs can be generated and run in parallel with the utility functions in this software package, without too much computational overhead.</p>
<p>Expected runtime: <em>phyloscanner</em> analysis of each batch of transmission networks takes about 2 hours, and the 345 batches can be processed in parallel.</p>
</div>
<div id="reconstructing-transmission-networks" class="section level3">
<h3 class="hasAnchor">
<a href="#reconstructing-transmission-networks" class="anchor"></a><a href="articles/Rakai.03.reconstruct_transmission_networks.html">Reconstructing transmission networks</a>
</h3>
<p><a href="articles/Rakai.03.reconstruct_transmission_networks.html">This tutorial</a> uses the generated <em>phyloscanner</em> output to reconstruct partially sampled HIV-1 transmission networks, of individuals who are in the population-based sample. We will illustrate a range of analytical functions to create these networks, and plotting functions to visualise phylogenetic inferences.</p>
<p>Expected runtime: 10 minutes.</p>
</div>
<div id="inferring-the-direction-of-transmission" class="section level3">
<h3 class="hasAnchor">
<a href="#inferring-the-direction-of-transmission" class="anchor"></a><a href="articles/Rakai.04.direction_of_transmission.html">Inferring the direction of transmission</a>
</h3>
<p><a href="articles/Rakai.04.direction_of_transmission.html">This tutorial</a> builds on the reconstructed HIV-1 transmission networks, and identifies phylogenetically highly supported source-recipient pairs. We will validate our phylogenetic inference into the direction of transmission against available epidemiologic and clinical data.</p>
<p>Expected runtime: 10 minutes.</p>
</div>
</div>
<div id="more-stuff" class="section level2">
<h2 class="hasAnchor">
<a href="#more-stuff" class="anchor"></a>More stuff</h2>
<ol>
<li>
<a href="articles/Stage1.create.read.alignments.html">Tutorial to generate read alignments for stage 1 analysis</a> .</li>
</ol>
</div>
<div id="testing" class="section level2">
<h2 class="hasAnchor">
<a href="#testing" class="anchor"></a>Testing</h2>
<p>The software has been run and tested on MacOS X 10.10, 10.11, 10.12 (Yosemite, El Capitan, Sierra) and CentOS Linux release 7.3.1611; and RAxML (8.2.9), MAFFT (7.212), samtools (1.2), anaconda (2.3.0), pysam (0.8.1), R (3.1), argparse (1.0.4), ape (4.1), data.table (1.10.5), dplyr (0.7.4), dtplyr (0.0.1), ff (2.2-13), ggplot2 (2.2.1.9000), grid (3.3.3), gridExtra (2.2.1), gtable (0.2.0), ggtree (1.6.9), kimisc (0.3), pegas (0.9), phangorn (2.1.1), phytools (0.5-64), prodlim (1.5.7), RColorBrewer (1.1-2), reshape (0.8.6), reshape2 (1.4.3), scales (0.5.0.9000), colorspace (1.3-2), devtools (1.13.4), ggnet (0.1.0), igraph (1.0.1), Rsamtools (1.26.1), RBGL (1.55.1), sna (2.4), knitr (1.17), rmarkdown (1.7).</p>
</div>
</div>

  </div>

  <div class="col-md-3" id="sidebar">
    <h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/olli0601/Phyloscanner.R.utilities">https://​github.com/​olli0601/​Phyloscanner.R.utilities</a>
</li>
<li>Report a bug at <br><a href="https://github.com/olli0601/Phyloscanner.R.utilities/issues">https://​github.com/​olli0601/​Phyloscanner.R.utilities/​issues</a>
</li>
</ul>
<h2>License</h2>
<p>GPL (&gt;=2)</p>
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Oliver Ratmann <br><small class="roles"> Author, maintainer </small> </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Oliver Ratmann.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
