<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Perform <em>phyloscanner</em> analysis on deep-sequence trees from a population-based sample • Phyloscanner.R.utilities</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">Phyloscanner.R.utilities</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Rakai.01.data_description.html">Data description</a>
    </li>
    <li>
      <a href="../articles/Rakai.01.run_phyloscanner.html">Perform *phyloscanner* analysis on deep-sequence trees from a population-based sample</a>
    </li>
    <li>
      <a href="../articles/Rakai.02.reconstruct_transmission_networks.html">Reconstructing transmission networks</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/olli0601/Phyloscanner.R.utilities">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Perform <em>phyloscanner</em> analysis on deep-sequence trees from a population-based sample</h1>
                        <h4 class="author">Oliver Ratmann</h4>
            
            <h4 class="date">2018-10-18</h4>
          </div>

    
    
<div class="contents">
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>This tutorial describes the steps to infer phylogenetic relationships from a large number of deep-sequence phylogenies of individuals in the same potential transmission network. The main objective is to illustrate how large numbers of phyloscanner runs can be generated and run in parallel with the utility functions in this software package, without too much computational overhead.</p>
<p>The tutorial assumes that <em>phyloscanner_make_trees.py</em> was already run to generate read alignments and deep-sequence phylogenies for individuals in the same potential transmission network. For the demo analysis of data from the Rakai population-based sample of 2,652 infected individuals, these data are provided in Data Set S1.</p>
</div>
<div id="setting-up-the-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-up-the-analysis" class="anchor"></a>Setting up the analysis</h2>
<p>Start by extracting Dataset S1 from the command line, assuming that it was copied into a data folder called ‘RakaiPopSample_data’:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">cd</span> /Users/Oliver/sandbox/DeepSeqProjects
<span class="kw">mkdir</span> RakaiPopSample_deepseqtrees
<span class="kw">tar</span> -xvf RakaiPopSample_data/Dataset_S1.tar -C RakaiPopSample_deepseqtrees</code></pre></div>
<p>The new directory should contain files that look as follows:</p>
<p align="center">
<img src="figures/Rakai.01.run_phyloscanner.directorystructure.png" alt="File structure of input directory"></p>
<ol style="list-style-type: decimal">
<li>Each analysis of a potential transmission network is identified with the prefix <code>ptyrX_</code> where <code>X</code> is an integer. Note that, to minimise computations, individuals in small transmission networks can be grouped into a single <em>phyloscanner</em> analysis, which we call a <em>batch</em> in this tutorial. In total, there are 345 batches.</li>
<li>Two files are available for each batch. First, a text file listing all individuals in the batch, in file <code>ptyrX_patients.txt</code>.</li>
<li>Second, deep-sequence phylogenies of reads from individuals in the same batch, generated from overlapping read alignments across the genome, in file <code>ptyrX_trees_newick.zip</code>. You can unzip the trees to see the coordinates of the genomic windows for which deep-sequence trees could be generated, but make sure you leave the zip files in the directory.</li>
</ol>
<p>Next, open R and define the base directories for your project:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(Phyloscanner.R.utilities)
HOME &lt;-<span class="st"> "~/sandbox/DeepSeqProjects"</span>
in.dir &lt;-<span class="st"> </span><span class="kw">file.path</span>(HOME, <span class="st">"RakaiPopSample_deepseqtrees"</span>)
out.dir &lt;-<span class="st"> </span><span class="kw">file.path</span>(HOME, <span class="st">"RakaiPopSample_phyloscanner_out"</span>)
work.dir &lt;-<span class="st"> </span><span class="kw">file.path</span>(HOME, <span class="st">"RakaiPopSample_phyloscanner_work"</span>)
prog.pty &lt;-<span class="st"> "/Users/Oliver/git/phylotypes/phyloscanner_make_trees.py"</span>
<span class="kw">dir.create</span>(out.dir, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)
<span class="kw">dir.create</span>(work.dir, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>Here, <code>prog.pty</code> is the full path to the <em>phyloscanner</em> program <em>phyloscanner_make_trees.py</em> and <code>work.dir</code> is the name of temporary directory.</p>
</div>
<div id="prepare-bash-scripts-to-run-phyloscanner" class="section level2">
<h2 class="hasAnchor">
<a href="#prepare-bash-scripts-to-run-phyloscanner" class="anchor"></a>Prepare bash scripts to run phyloscanner</h2>
<p>The next step is to define the input arguments to <em>phyloscanner</em>. Please see the <em>phyloscanner</em> manual for details. The default arguments that were used for analysis of the Rakai population-based sample are as follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pty.args &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">prog.pty =</span> prog.pty, <span class="dt">prog.mafft =</span> <span class="ot">NA</span>, <span class="dt">prog.raxml =</span> <span class="ot">NA</span>, <span class="dt">data.dir =</span> <span class="ot">NA</span>, 
    <span class="dt">work.dir =</span> work.dir, <span class="dt">out.dir =</span> out.dir, <span class="dt">alignments.file =</span> <span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">"Phyloscanner.R.utilities"</span>, 
        <span class="st">"HIV1_compendium_AD_B_CPX_v2.fasta"</span>), <span class="dt">alignments.root =</span> <span class="st">"REF_CPX_AF460972"</span>, 
    <span class="dt">alignments.pairwise.to =</span> <span class="st">"REF_B_K03455"</span>, <span class="dt">bl.normalising.reference.file =</span> <span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">"Phyloscanner.R.utilities"</span>, 
        <span class="st">"data"</span>, <span class="st">"hiv.hxb2.norm.constants.rda"</span>), <span class="dt">bl.normalising.reference.var =</span> <span class="st">"MEDIAN_PWD"</span>, 
    <span class="dt">window.automatic =</span> <span class="st">""</span>, <span class="dt">merge.threshold =</span> <span class="dv">0</span>, <span class="dt">min.read.count =</span> <span class="dv">1</span>, <span class="dt">quality.trim.ends =</span> <span class="dv">23</span>, 
    <span class="dt">min.internal.quality =</span> <span class="dv">23</span>, <span class="dt">merge.paired.reads =</span> <span class="ot">TRUE</span>, <span class="dt">no.trees =</span> <span class="ot">FALSE</span>, 
    <span class="dt">dont.check.duplicates =</span> <span class="ot">FALSE</span>, <span class="dt">dont.check.recombination =</span> <span class="ot">TRUE</span>, <span class="dt">num.bootstraps =</span> <span class="dv">1</span>, 
    <span class="dt">all.bootstrap.trees =</span> <span class="ot">TRUE</span>, <span class="dt">strip.max.len =</span> <span class="dv">350</span>, <span class="dt">min.ureads.individual =</span> <span class="ot">NA</span>, 
    <span class="dt">win =</span> <span class="kw">c</span>(<span class="dv">800</span>, <span class="dv">9400</span>, <span class="dv">25</span>, <span class="dv">250</span>), <span class="dt">keep.overhangs =</span> <span class="ot">FALSE</span>, <span class="dt">use.blacklisters =</span> <span class="kw">c</span>(<span class="st">"ParsimonyBasedBlacklister"</span>, 
        <span class="st">"DownsampleReads"</span>), <span class="dt">tip.regex =</span> <span class="st">"^(.*)_fq[0-9]+_read_([0-9]+)_count_([0-9]+)$"</span>, 
    <span class="dt">roguesubtree.kParam =</span> <span class="dv">20</span>, <span class="dt">roguesubtree.prop.threshold =</span> <span class="dv">0</span>, <span class="dt">roguesubtree.read.threshold =</span> <span class="dv">20</span>, 
    <span class="dt">dwns.maxReadsPerPatient =</span> <span class="dv">50</span>, <span class="dt">multifurcation.threshold =</span> <span class="fl">1e-05</span>, <span class="dt">split.rule =</span> <span class="st">"s"</span>, 
    <span class="dt">split.kParam =</span> <span class="dv">20</span>, <span class="dt">split.proximityThreshold =</span> <span class="dv">0</span>, <span class="dt">split.readCountsMatterOnZeroBranches =</span> <span class="ot">TRUE</span>, 
    <span class="dt">split.pruneBlacklist =</span> <span class="ot">FALSE</span>, <span class="dt">trms.allowMultiTrans =</span> <span class="ot">TRUE</span>, <span class="dt">pw.trmw.min.reads =</span> <span class="dv">30</span>, 
    <span class="dt">pw.trmw.min.tips =</span> <span class="dv">1</span>, <span class="dt">pw.trmw.close.brl =</span> <span class="fl">0.025</span>, <span class="dt">pw.trmw.distant.brl =</span> <span class="fl">0.05</span>, 
    <span class="dt">pw.prior.keff =</span> <span class="dv">2</span>, <span class="dt">pw.prior.neff =</span> <span class="dv">3</span>, <span class="dt">pw.prior.keff.dir =</span> <span class="dv">2</span>, <span class="dt">pw.prior.neff.dir =</span> <span class="dv">3</span>, 
    <span class="dt">pw.prior.calibrated.prob =</span> <span class="fl">0.66</span>, <span class="dt">mem.save =</span> <span class="dv">0</span>, <span class="dt">verbose =</span> <span class="ot">TRUE</span>, <span class="dt">select =</span> <span class="ot">NA</span>)
<span class="kw">save</span>(pty.args, <span class="dt">file =</span> <span class="kw">file.path</span>(out.dir, <span class="st">"pty.args.rda"</span>))</code></pre></div>
<p>Next, we will prepare UNIX <em>bash</em> scripts to run a large number of <em>phyloscanner</em> analyses on a population-based sample. Each <em>bash</em> script corresponds to the deep-sequence phylogenetic analysis of one batch of individuals that are analysed jointly.</p>
<p>For each batch of individuals to be processed, find the corresponding list of patients in the input directory:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pty.c &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">FILE_PAT =</span> <span class="kw">list.files</span>(in.dir, <span class="dt">pattern =</span> <span class="st">"_patients.txt"</span>, 
    <span class="dt">full.names =</span> <span class="ot">TRUE</span>))
pty.c[, <span class="st">`</span><span class="dt">:=</span><span class="st">`</span>(PTY_RUN, <span class="kw">as.integer</span>(<span class="kw">gsub</span>(<span class="st">"ptyr"</span>, <span class="st">""</span>, <span class="kw">gsub</span>(<span class="st">"_patients.txt"</span>, <span class="st">""</span>, 
    <span class="kw">basename</span>(FILE_PAT)))))]</code></pre></div>
<p>Check which (if any) batches have already been processed, and remove them from the TODO list:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tmp &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">FILE_TRMW =</span> <span class="kw">list.files</span>(out.dir, <span class="dt">pattern =</span> <span class="st">"_pairwise_relationships.rda"</span>, 
    <span class="dt">full.names =</span> <span class="ot">TRUE</span>))
tmp[, <span class="st">`</span><span class="dt">:=</span><span class="st">`</span>(PTY_RUN, <span class="kw">as.integer</span>(<span class="kw">gsub</span>(<span class="st">"ptyr"</span>, <span class="st">""</span>, <span class="kw">gsub</span>(<span class="st">"_pairwise_relationships.rda"</span>, 
    <span class="st">""</span>, <span class="kw">basename</span>(FILE_TRMW)))))]
pty.c &lt;-<span class="st"> </span><span class="kw">merge</span>(pty.c, tmp, <span class="dt">by =</span> <span class="st">"PTY_RUN"</span>, <span class="dt">all.x =</span> <span class="dv">1</span>)
pty.c &lt;-<span class="st"> </span><span class="kw">subset</span>(pty.c, <span class="kw">is.na</span>(FILE_TRMW))</code></pre></div>
<p>For each batch of individuals, create a UNIX <em>bash</em> script to run <em>phyloscanner</em>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">setkey</span>(pty.c, PTY_RUN)
pty.c &lt;-<span class="st"> </span>pty.c[, {
    prefix.infiles &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">"patients.txt"</span>, <span class="st">""</span>, FILE_PAT)
    cmd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/phsc.cmd.phyloscanner.one.resume.html">phsc.cmd.phyloscanner.one.resume</a></span>(prefix.infiles, pty.args)
    <span class="kw">list</span>(<span class="dt">CMD =</span> cmd)
}, by =<span class="st"> "PTY_RUN"</span>]</code></pre></div>
<p>Each bash script should look similar to this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pty.c[<span class="dv">1</span>, <span class="kw">cat</span>(CMD)]</code></pre></div>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ot">CWD=$(</span><span class="kw">pwd</span><span class="ot">)</span>
<span class="kw">echo</span> <span class="ot">$CWD</span>
<span class="kw">mkdir</span> -p <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41"</span>
<span class="kw">cp</span> <span class="st">"/Users/Oliver/sandbox/DeepSeqProjects/RakaiPopSample_deepseqtrees/ptyr2_patients.txt"</span> <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41"</span>
<span class="kw">unzip</span> <span class="st">"/Users/Oliver/sandbox/DeepSeqProjects/RakaiPopSample_deepseqtrees/ptyr2_trees_fasta.zip"</span> -d <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41"</span>
<span class="kw">unzip</span> <span class="st">"/Users/Oliver/sandbox/DeepSeqProjects/RakaiPopSample_deepseqtrees/ptyr2_trees_newick.zip"</span> -d <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41"</span>
<span class="kw">cd</span> <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41"</span>
<span class="kw">Rscript</span> /Users/Oliver/git/phylotypes/deprecated/NormalisationLookupWriter.R --scriptdir /Users/Oliver/git/phylotypes/deprecated <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_InWindow_"</span> <span class="st">"/Users/Oliver/Library/R/3.3/library/Phyloscanner.R.utilities/data/hiv.hxb2.norm.constants.rda"</span> <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_normconst.csv"</span> <span class="st">"MEDIAN_PWD"</span>  --standardize
<span class="kw">Rscript</span> /Users/Oliver/git/phylotypes/tools/parsimony_based_blacklister.R 20 0 20 <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_InWindow_"</span> <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_blacklistsank_InWindow"</span> --dualsOutputFile <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_duallistsank_InWindow"</span> --outgroupName REF_CPX_AF460972 --tipRegex <span class="st">"^(.*)_fq[0-9]+_read_([0-9]+)_count_([0-9]+)$"</span> --multifurcationThreshold 1e-05  --branchLengthNormalisation <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_normconst.csv"</span> --verbose
<span class="kw">Rscript</span> /Users/Oliver/git/phylotypes/tools/downsample_reads.R 50 <span class="ot">$CWD</span>/pty_18-10-19-15-43-41/ptyr2_ <span class="ot">$CWD</span>/pty_18-10-19-15-43-41/ptyr2_blacklistdwns_ --blacklist <span class="ot">$CWD</span>/pty_18-10-19-15-43-41/ptyr2_blacklistsank_InWindow_ --tipRegex <span class="st">"^(.*)_fq[0-9]+_read_([0-9]+)_count_([0-9]+)$"</span> --seed 42 --verbose
<span class="kw">Rscript</span> /Users/Oliver/git/phylotypes/tools/split_hosts_to_subgraphs.R <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_"</span> <span class="st">"ptyr2"</span> --blacklist <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_blacklistdwns_"</span> --outputdir <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41"</span> --idFile <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_patients.txt"</span> --outgroupName REF_CPX_AF460972 --splitsRule s --kParam 20 --proximityThreshold 0 --readCountsMatterOnZeroBranches --tipRegex <span class="st">"^(.*)_fq[0-9]+_read_([0-9]+)_count_([0-9]+)$"</span> --multifurcationThreshold 1e-05 --branchLengthNormalisation <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_normconst.csv"</span> --outputAsRDA --pdfwidth 30 --pdfrelheight 0.15 --verbose
<span class="kw">Rscript</span> /Users/Oliver/git/phylotypes/tools/classify_relationships.R <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ProcessedTree_s_ptyr2_"</span> <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/subgraphs_s_ptyr2_"</span> <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2"</span> --branchLengthNormalisation <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_normconst.csv"</span> --verbose
<span class="kw">Rscript</span> /Users/Oliver/git/phylotypes/tools/summary_statistics.R --scriptDir /Users/Oliver/git/phylotypes/tools <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_patients.txt"</span> <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ProcessedTree_s_ptyr2_InWindow_"</span> <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/subgraphs_s_ptyr2_InWindow_"</span> <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_"</span> --tipRegex <span class="st">"^(.*)_fq[0-9]+_read_([0-9]+)_count_([0-9]+)$"</span> --blacklists <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_blacklistdwns_InWindow_"</span> --verbose
<span class="kw">Rscript</span> /Users/Oliver/git/phylotypes/deprecated/TransmissionSummary.R <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_patients.txt"</span> <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_classification_InWindow_"</span> <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_trmStats.csv"</span> --scriptdir /Users/Oliver/git/phylotypes/deprecated --summaryFile <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_patStatsFull.csv"</span> --minThreshold 1 --detailedOutput <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_trmStatsPerWindow.rda"</span> --allowMultiTrans --verbose
<span class="kw">Rscript</span> /Users/Oliver/Library/R/3.3/library/Phyloscanner.R.utilities/phsc.pairwise.relationships.Rscript --infile <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_trmStatsPerWindow.rda"</span> --outfile <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_pairwise_relationships.rda"</span> --trmw.min.reads 30 --trmw.min.tips 1 --trmw.close.brl 0.025 --trmw.distant.brl 0.05 --prior.keff 2 --prior.neff 3 --prior.keff.dir 2 --prior.neff.dir 3 --prior.calibrated.prob 0.66 --rel.likely.pair --rel.likely.pair.by.distance.only --rel.likely.pair.by.topology.only --rel.likely.pair.by.cross.table --rel.direction --rel.chain
<span class="kw">Rscript</span> /Users/Oliver/Library/R/3.3/library/Phyloscanner.R.utilities/phsc.read.processed.phyloscanner.output.in.directory.Rscript --prefix.infiles <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_"</span> --save.file.base <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41/ptyr2_"</span> --read.likelytransmissions --read.trees --read.subtrees --zip
<span class="kw">mv</span> ptyr2* <span class="st">"~/sandbox/DeepSeqProjects/RakaiPopSample_phyloscanner_out"</span>
<span class="kw">for</span> <span class="kw">file</span> in *<span class="kw">;</span> <span class="kw">do</span>
    <span class="kw">zip</span> -ur9XTj ptyr2_otherstuff.zip <span class="st">"</span><span class="ot">$file</span><span class="st">"</span>
<span class="kw">done</span>
<span class="kw">mv</span> ptyr2_otherstuff.zip <span class="st">"~/sandbox/DeepSeqProjects/RakaiPopSample_phyloscanner_out"</span>
<span class="kw">cd</span> <span class="ot">$CWD</span>
<span class="kw">rm</span> -r <span class="st">"</span><span class="ot">$CWD</span><span class="st">/pty_18-10-19-15-43-41"</span>  </code></pre></div>
</div>
<div id="run-bash-scripts-option-1" class="section level2">
<h2 class="hasAnchor">
<a href="#run-bash-scripts-option-1" class="anchor"></a>Run bash scripts (option 1)</h2>
<p>Each <em>bash</em> script can be run from a UNIX terminal, we only have to write the scripts to file. This following code will write each script to the temp directory that you specified above, to a file named <code>phsc.Wed_Oct_17_101858_2018.sh</code> or similar:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">invisible</span>(pty.c[, {
    outfile &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">":"</span>, <span class="st">""</span>, <span class="kw">paste</span>(<span class="st">"phsc"</span>, <span class="kw">paste</span>(<span class="kw">strsplit</span>(<span class="kw">date</span>(), <span class="dt">split =</span> <span class="st">" "</span>)[[<span class="dv">1</span>]], 
        <span class="dt">collapse =</span> <span class="st">"_"</span>, <span class="dt">sep =</span> <span class="st">""</span>), <span class="st">"sh"</span>, <span class="dt">sep =</span> <span class="st">"."</span>))
    outfile &lt;-<span class="st"> </span><span class="kw">file.path</span>(pty.args[[<span class="st">"work.dir"</span>]], outfile)
    <span class="kw">cat</span>(CMD, <span class="dt">file =</span> outfile)
    <span class="kw">Sys.chmod</span>(outfile, <span class="dt">mode =</span> <span class="st">"777"</span>)
    <span class="kw">Sys.sleep</span>(<span class="dv">1</span>)
}, <span class="dt">by =</span> <span class="st">"PTY_RUN"</span>])</code></pre></div>
<p>Each file can be run on a UNIX terminal, e.g.:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cd /Users/Oliver/<span class="kw">Dropbox</span> (SPH Imperial College)/2015_PANGEA_DualPairsFromFastQIVA/RakaiPopSample_phyloscanner_work
phsc.Wed_Oct_17_101858_2018.sh  </code></pre></div>
</div>
<div id="run-bash-scripts-option-2" class="section level2">
<h2 class="hasAnchor">
<a href="#run-bash-scripts-option-2" class="anchor"></a>Run bash scripts (option 2)</h2>
<p>Alternatively, the bash scripts can be processed on a high performance environment with a job scheduling system. We first define a PBS header for the job scheduling system, add the header to each script, and then submit each script to the job scheduling system. The exact form of the PBS header depends on your job scheduler, below is an example that works at Imperial.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hpc.load &lt;-<span class="st"> "module load R/3.3.3"</span>  <span class="co"># make R available </span>
hpc.select &lt;-<span class="st"> </span><span class="dv">1</span>  <span class="co"># number of nodes</span>
hpc.nproc &lt;-<span class="st"> </span><span class="dv">1</span>  <span class="co"># number of processors on node</span>
hpc.walltime &lt;-<span class="st"> </span><span class="dv">15</span>  <span class="co"># walltime</span>
hpc.q &lt;-<span class="st"> "pqeelab"</span>  <span class="co"># PBS queue</span>
hpc.mem &lt;-<span class="st"> "6gb"</span>  <span class="co"># RAM</span>
pbshead &lt;-<span class="st"> "#!/bin/sh"</span>
tmp &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">"#PBS -l walltime="</span>, hpc.walltime, <span class="st">":59:59,pcput="</span>, hpc.walltime, 
    <span class="st">":45:00"</span>, <span class="dt">sep =</span> <span class="st">""</span>)
pbshead &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, tmp, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
tmp &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">"#PBS -l select="</span>, hpc.select, <span class="st">":ncpus="</span>, hpc.nproc, <span class="st">":mem="</span>, hpc.mem, 
    <span class="dt">sep =</span> <span class="st">""</span>)
pbshead &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, tmp, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
pbshead &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, <span class="st">"#PBS -j oe"</span>, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
pbshead &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, <span class="kw">paste</span>(<span class="st">"#PBS -q"</span>, hpc.q), <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
pbshead &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, hpc.load, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</code></pre></div>
<p>Our header thus looks as follows:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">#!/bin/sh </span>
<span class="co">#PBS -l walltime=15:59:59,pcput=15:45:00 </span>
<span class="co">#PBS -l select=1:ncpus=1:mem=6gb </span>
<span class="co">#PBS -j oe </span>
<span class="co">#PBS -q pqeelab </span>
<span class="kw">module</span> load R/3.3.3</code></pre></div>
<p>We are now ready to add the header to each script, and submit the job:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">invisible</span>(pty.c[, {
    cmd &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, <span class="st">"cd $TMPDIR"</span>, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
    cmd &lt;-<span class="st"> </span><span class="kw">paste</span>(cmd, CMD, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
    outfile &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">":"</span>, <span class="st">""</span>, <span class="kw">paste</span>(<span class="st">"phsc"</span>, <span class="kw">paste</span>(<span class="kw">strsplit</span>(<span class="kw">date</span>(), <span class="dt">split =</span> <span class="st">" "</span>)[[<span class="dv">1</span>]], 
        <span class="dt">collapse =</span> <span class="st">"_"</span>, <span class="dt">sep =</span> <span class="st">""</span>), <span class="st">"sh"</span>, <span class="dt">sep =</span> <span class="st">"."</span>))
    outfile &lt;-<span class="st"> </span><span class="kw">file.path</span>(pty.args[[<span class="st">"work.dir"</span>]], outfile)
    <span class="kw">cat</span>(CMD, <span class="dt">file =</span> outfile)
    cmd &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">"qsub"</span>, outfile)
    <span class="kw">cat</span>(cmd)
    <span class="kw">cat</span>(<span class="kw">system</span>(cmd, <span class="dt">intern =</span> <span class="ot">TRUE</span>))
    <span class="kw">Sys.sleep</span>(<span class="dv">1</span>)
}, <span class="dt">by =</span> <span class="st">"PTY_RUN"</span>])</code></pre></div>
</div>
<div id="expected-phyloscanner-output" class="section level2">
<h2 class="hasAnchor">
<a href="#expected-phyloscanner-output" class="anchor"></a>Expected <em>phyloscanner</em> output</h2>
<p>Once all scripts are run, the output directory contains for each batch a number of files, of which the files called <code>ptyrX_pairwise_relationships.rda</code> are used for reconstructing HIV-1 transmission networks from the population-based sample.</p>
<p>Below is a description of the full output.</p>
<ol style="list-style-type: decimal">
<li>
<code>ptyrX_patients.txt</code> Input file, list of individuals in this batch.</li>
<li>
<code>ptyrX_trees_newick.zip</code> Input file, deep sequence trees.</li>
<li>
<code>ptyrX_pairwise_relationships.rda</code> Main output file for reconstructing HIV-1 transmission networks.<br>
</li>
<li>
<code>ptyrX_normconst.csv</code> File containing multipliers used to standardise branch lengths of each deep-sequence phylogeny.</li>
<li>
<code>ptyrX_trees.rda</code> Deep-sequence phylogenies in <em>ape</em> format, annotated with subgraphs of each individual.</li>
<li>
<code>ptyrX_trees_processed_nexus.zip</code> Deep-sequence phylogenies in <em>nexus</em> format, annotated with subgraphs of each individual.</li>
<li>
<code>ptyrX_trees_collapsed.zip</code> Collaposed deep-sequence phylogenies, with all blacklisted taxa removed.</li>
<li>
<code>ptyrX_trees_pdf.zip</code> PDFs of the deep-sequence phylogenies.</li>
<li>
<code>ptyrX_patStatsFull.csv</code> Detailed description of subgraphs of all individuals in this batch across the genome.</li>
<li>
<code>ptyrX_patStatsSummary.csv</code> Summary description of subgraphs of all individuals in this batch across the genome.</li>
<li>
<code>ptyrX_patStats.pdf</code> PDFs visualising properties of the subgraphs of all individuals in this batch across the genome.</li>
<li>
<code>ptyrX_trmStats.zip</code> Detailed description of pairwise phylogenetic relationships of all individuals in this batch across the genome, in <em>csv</em> format.</li>
<li>
<code>ptyrX_trmStatsPerWindow.rda</code> Detailed description of pairwise phylogenetic relationships of all individuals in this batch across the genome, in <em>R</em> format.</li>
<li>
<code>ptyrX_trmStats.csv</code> Summary description of pairwise phylogenetic relationships of all individuals in this batch across the genome.</li>
<li>
<code>ptyrX_subtrees_r_csv.zip</code> Detailed description of subgraphs of all individuals in this batch across the genome, in <em>csv</em> format.</li>
<li>
<code>ptyrX_subtrees_r_rda.zip</code> Detailed description of subgraphs of all individuals in this batch across the genome, in <em>R</em> format.<br>
</li>
<li>
<code>ptyrX_trees_blacklist.zip</code> Names of blacklisted taxa of all individuals in this batch.</li>
<li>
<code>ptyrX_trees_duallist.zip</code> Names of potential contaminants of all individuals in this batch.</li>
</ol>
Once all scripts are run, the output directory should have a file structure similar to that shown here:
<p align="center">
<img src="figures/Rakai.01.run_phyloscanner.expectedoutput.png" alt="Expected output of phyloscanner analysis of potential transmission networks"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#setting-up-the-analysis">Setting up the analysis</a></li>
      <li><a href="#prepare-bash-scripts-to-run-phyloscanner">Prepare bash scripts to run phyloscanner</a></li>
      <li><a href="#run-bash-scripts-option-1">Run bash scripts (option 1)</a></li>
      <li><a href="#run-bash-scripts-option-2">Run bash scripts (option 2)</a></li>
      <li><a href="#expected-phyloscanner-output">Expected <em>phyloscanner</em> output</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Oliver Ratmann.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
