<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Stage 1 analysis: identify phylogenetically closely related individuals • Phyloscanner.R.utilities</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Stage 1 analysis: identify phylogenetically closely related individuals">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Phyloscanner.R.utilities</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.7.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Rakai.01.data_description.html">Data description</a>
    </li>
    <li>
      <a href="../articles/Rakai.02.run_phyloscanner.html">Running *phyloscanner* in high-throughput on a population-based sample</a>
    </li>
    <li>
      <a href="../articles/Rakai.03.reconstruct_transmission_networks.html">Reconstructing transmission networks</a>
    </li>
    <li>
      <a href="../articles/Rakai.04.direction_of_transmission.html">Inferring the direction of transmission</a>
    </li>
    <li>
      <a href="../articles/Stage1.create.read.alignments.html">Stage 1 analysis: identify phylogenetically closely related individuals</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/olli0601/Phyloscanner.R.utilities">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Stage 1 analysis: identify phylogenetically closely related individuals</h1>
                        <h4 class="author">Oliver Ratmann</h4>
            
            <h4 class="date">2018-10-30</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/olli0601/Phyloscanner.R.utilities/blob/master/vignettes/Stage1.create.read.alignments.Rmd"><code>vignettes/Stage1.create.read.alignments.Rmd</code></a></small>
      <div class="hidden name"><code>Stage1.create.read.alignments.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>This tutorial describes how to identify phylogenetically closely related individuals with <em>phyloscanner</em> from of a large population-based sample of deep sequences.</p>
</div>
<div id="setting-up-the-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-up-the-analysis" class="anchor"></a>Setting up the analysis</h2>
<p>Let us define input and output directories for our analysis in R.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(Phyloscanner.R.utilities)
HOME &lt;&lt;-<span class="st"> "/rds/general/project/ratmann_pangea_analyses_mrc_uvri/live"</span>
data.dir &lt;-<span class="st"> "/rds/general/project/ratmann_pangea_deepsequencedata/live/PANGEA2_MRC"</span>
prog.pty &lt;-<span class="st"> "/rds/general/user/or105/home/phyloscanner/phyloscanner_make_trees.py"</span>
in.dir &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(HOME, <span class="st">"MRCPopSample_phsc_stage1_input"</span>)
work.dir &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(HOME, <span class="st">"MRCPopSample_phsc_work"</span>)
out.dir &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(HOME, <span class="st">"MRCPopSample_phsc_stage1_output"</span>)
<span class="kw"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span>(in.dir, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)
<span class="kw"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span>(out.dir, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)
<span class="kw"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span>(work.dir, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>Here, <code>prog.pty</code> is the full path to the <em>phyloscanner</em> program <em>phyloscanner_make_trees.py</em> and <code>work.dir</code> is the name of a temporary directory. Do make sure that the directory names above do not start with “~”, because the names are not expanded in the scripts below. White space, or characters like ‘-’ are OK.</p>
</div>
<div id="setting-up-phyloscanner-runs-to-test-the-phylogenetic-relationship-of-all-possible-pairs-of-individuals" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-up-phyloscanner-runs-to-test-the-phylogenetic-relationship-of-all-possible-pairs-of-individuals" class="anchor"></a>Setting up <em>phyloscanner</em> runs to test the phylogenetic relationship of all possible pairs of individuals</h2>
<p>Next, we define a large number of <em>phyloscanner</em> runs to test the phylogenetic relationship of all possible pairs of individuals. We start by generating a list of all bam files in our data directory (<code>data.dir</code>), and then we identify the corresponding individuals with a regular expression:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dbam &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">BAM =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(data.dir, <span class="dt">pattern =</span> <span class="st">".*remap</span><span class="ch">\\</span><span class="st">.bam$"</span>, <span class="dt">full.names =</span> <span class="ot">FALSE</span>, 
    <span class="dt">recursive =</span> <span class="ot">TRUE</span>))
regex.person &lt;-<span class="st"> "^([A-Z0-9]+-[A-Z0-9]+)-.*$"</span>
dbam[, <span class="st">`</span><span class="dt">:=</span><span class="st">`</span>(IND, <span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(regex.person, <span class="st">"</span><span class="ch">\\</span><span class="st">1"</span>, BAM))]</code></pre></div>
The data.table <code>dbam</code> should look similar to the following:
<p align="center">
<img src="figures/Stage1.bams.png" alt="List of bam files, stage 1 of analysis of large population-based sample."></p>
<p>We then group individuals for phyloscanner analyses, so that phylogenetic linkage between every pair of individuals is assessed at least once. Specifically, individuals are grouped into batches of specified size, and then, all possible pairs of batches are formed. Each of these pairs of batches defines a group of individuals between whom phylogenetic linkages are assessed in one phyloscanner run. Thus, the number of individuals in each group is twice the batch size:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tmp &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(dbam$IND)
pty.runs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/phsc.define.stage1.analyses.html">phsc.define.stage1.analyses</a></span>(tmp, <span class="dt">batch.size =</span> <span class="dv">50</span>)
<span class="co"># add BAM files per individual</span>
pty.runs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/merge.html">merge</a></span>(pty.runs, <span class="kw"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(dbam, <span class="dt">select =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(IND, BAM)), <span class="dt">by =</span> <span class="st">"IND"</span>, 
    <span class="dt">allow.cartesian =</span> <span class="ot">TRUE</span>)
<span class="kw">setkey</span>(pty.runs, PTY_RUN)
<span class="co"># save to file</span>
outfile &lt;-<span class="st"> "phsc_runs_MRC_stage1_n2531_181026.csv"</span>
<span class="kw"><a href="https://rdrr.io/r/utils/write.table.html">write.csv</a></span>(pty.runs, <span class="dt">file =</span> <span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(in.dir, outfile), <span class="dt">row.names =</span> <span class="ot">FALSE</span>)</code></pre></div>
The data.table <code>pty.runs</code> should look similar to the following:
<p align="center">
<img src="figures/Stage1.predefinedruns.png" alt="List of phyloscanner runs, stage 1 of analysis of large population-based sample."></p>
<p>The data.table specifies a large number of <em>phyloscanner</em> runs. All individuals (‘IND’) are grouped in batches (‘BATCH’), and two batches are paired to evaluate the phylogenetic relationship of the associated individuals in a particular analysis (‘PTY_RUN’).</p>
</div>
<div id="prepare-bash-scripts-to-generate-read-alignments" class="section level2">
<h2 class="hasAnchor">
<a href="#prepare-bash-scripts-to-generate-read-alignments" class="anchor"></a>Prepare bash scripts to generate read alignments</h2>
<p>The next step is to define the input arguments generate read alignments with <em>phyloscanner</em>. Please see the <em>phyloscanner</em> manual for details. The default arguments that were used for analysis of the Rakai population-based sample are as follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pty.select &lt;-<span class="st"> </span><span class="dv">1</span>:<span class="dv">2</span>  <span class="co">#for demo purposes, we select the first two runs. Set to NA.</span>
pty.args &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">prog.pty =</span> prog.pty, <span class="dt">prog.mafft =</span> <span class="st">"mafft"</span>, <span class="dt">data.dir =</span> data.dir, 
    <span class="dt">work.dir =</span> work.dir, <span class="dt">out.dir =</span> out.dir, <span class="dt">alignments.file =</span> <span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="dt">package =</span> <span class="st">"Phyloscanner.R.utilities"</span>, 
        <span class="st">"HIV1_compendium_AD_B_CPX_v2.fasta"</span>), <span class="dt">alignments.root =</span> <span class="st">"REF_CPX_AF460972"</span>, 
    <span class="dt">alignments.pairwise.to =</span> <span class="st">"REF_B_K03455"</span>, <span class="dt">window.automatic =</span> <span class="st">""</span>, <span class="dt">merge.threshold =</span> <span class="dv">2</span>, 
    <span class="dt">min.read.count =</span> <span class="dv">1</span>, <span class="dt">quality.trim.ends =</span> <span class="dv">23</span>, <span class="dt">min.internal.quality =</span> <span class="dv">23</span>, <span class="dt">merge.paired.reads =</span> <span class="ot">TRUE</span>, 
    <span class="dt">no.trees =</span> <span class="ot">TRUE</span>, <span class="dt">dont.check.duplicates =</span> <span class="ot">FALSE</span>, <span class="dt">dont.check.recombination =</span> <span class="ot">TRUE</span>, 
    <span class="dt">win =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">800</span>, <span class="dv">9400</span>, <span class="dv">25</span>, <span class="dv">250</span>), <span class="dt">keep.overhangs =</span> <span class="ot">FALSE</span>, <span class="dt">mem.save =</span> <span class="dv">0</span>, <span class="dt">verbose =</span> <span class="ot">TRUE</span>, 
    <span class="dt">select =</span> pty.select  <span class="co">#of 240</span>
)
<span class="kw"><a href="https://rdrr.io/r/base/save.html">save</a></span>(pty.args, <span class="dt">file =</span> <span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(in.dir, <span class="st">"phsc_args_stage1_create_read_alignments.rda"</span>))</code></pre></div>
<p>Next, we will prepare UNIX bash scripts to process the data in parallel. Each bash script will consist of UNIX commands to generate alignments of deep-sequence viral reads for one phyloscanner run (‘PTY_RUN’). All scripts, and hence all runs can be performed in parallel. Each alignment contains viral reads of the individuals in one run, and all reads overlap a particular genomic window of the HIV-1 genome:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># check which (if any) batches have already been processed, and remove from</span>
<span class="co"># TODO list</span>
tmp &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">FILE_FASTA =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(out.dir, <span class="dt">pattern =</span> <span class="st">"^ptyr[0-9]+_"</span>, 
    <span class="dt">full.names =</span> <span class="ot">TRUE</span>))
tmp[, <span class="st">`</span><span class="dt">:=</span><span class="st">`</span>(PTY_RUN, <span class="kw"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"ptyr([0-9]+)_.*"</span>, <span class="st">"</span><span class="ch">\\</span><span class="st">1"</span>, <span class="kw"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(FILE_FASTA))))]
pty.runs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/merge.html">merge</a></span>(pty.runs, tmp, <span class="dt">by =</span> <span class="st">"PTY_RUN"</span>, <span class="dt">all.x =</span> <span class="dv">1</span>)
pty.runs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(pty.runs, <span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(FILE_FASTA))
<span class="co"># search for bam files and references and merge with runs</span>
pty.runs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(pty.runs, <span class="dt">select =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(PTY_RUN, IND))
tmp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/phsc.find.bam.and.references.html">phsc.find.bam.and.references</a></span>(pty.args[[<span class="st">"data.dir"</span>]], <span class="dt">regex.person =</span> <span class="st">"^([A-Z0-9]+-[A-Z0-9]+)-.*$"</span>)
pty.runs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/merge.html">merge</a></span>(pty.runs, tmp, <span class="dt">by =</span> <span class="st">"IND"</span>)
<span class="co"># create UNIX bash scripts</span>
pty.runs &lt;-<span class="st"> </span>pty.runs[, <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">BAM =</span> BAM, <span class="dt">REF =</span> REF, <span class="dt">SAMPLE =</span> SAMPLE, <span class="dt">RENAME_ID =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(IND, 
    <span class="st">"-fq"</span>, <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(BAM)))), by =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"PTY_RUN"</span>, <span class="st">"IND"</span>)]
<span class="kw">setkey</span>(pty.runs, PTY_RUN)
<span class="kw">setnames</span>(pty.runs, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"IND"</span>, <span class="st">"SAMPLE"</span>), <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"UNIT_ID"</span>, <span class="st">"SAMPLE_ID"</span>))
pty.c &lt;-<span class="st"> </span><span class="kw"><a href="../reference/phsc.cmd.phyloscanner.multi.html">phsc.cmd.phyloscanner.multi</a></span>(pty.runs, pty.args)
pty.c[, <span class="st">`</span><span class="dt">:=</span><span class="st">`</span>(CASE_ID, <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(pty.c)))]</code></pre></div>
<p>Let us take a look at the first batch script:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pty.c[<span class="dv">1</span>, <span class="kw"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(CMD)]</code></pre></div>
<p align="center">
<img src="figures/Stage1.createbamscmd.png" alt="Bash script to generate read alignments, stage 1 of analysis of large population-based sample."></p>
<p>We see the familiar call to <code>phyloscanner_make_trees.py</code>, and we suppressed tree reconstruction with the input command <code>--no-trees</code>. There is also a bit of file copying before/after. This ensures that the script is self-contained, and that it operates on a local directory in case the data are stored on a central data storage system.</p>
</div>
<div id="run-bash-scripts-to-generate-read-alignments" class="section level2">
<h2 class="hasAnchor">
<a href="#run-bash-scripts-to-generate-read-alignments" class="anchor"></a>Run bash scripts to generate read alignments</h2>
<p>The next task is submit the bash scripts to a job scheduling system on a high performance computing environment. We first define a PBS header for the job scheduling system, add the header to each script, and then submit each script to the job scheduling system. The exact form of the PBS header depends on your job scheduler, below is an example that works at Imperial.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hpc.load &lt;-<span class="st"> "module load R/3.3.3"</span>  <span class="co"># make R available </span>
hpc.select &lt;-<span class="st"> </span><span class="dv">1</span>  <span class="co"># number of nodes</span>
hpc.nproc &lt;-<span class="st"> </span><span class="dv">1</span>  <span class="co"># number of processors on node</span>
hpc.walltime &lt;-<span class="st"> </span><span class="dv">15</span>  <span class="co"># walltime</span>
hpc.q &lt;-<span class="st"> "pqeelab"</span>  <span class="co"># PBS queue</span>
hpc.mem &lt;-<span class="st"> "6gb"</span>  <span class="co"># RAM</span>
hpc.array &lt;-<span class="st"> </span>pty.c[, <span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(CASE_ID)]  <span class="co"># number of runs for job array\t</span>
<span class="co"># define PBS header for job scheduler. this will depend on your job</span>
<span class="co"># scheduler.</span>
pbshead &lt;-<span class="st"> "#!/bin/sh"</span>
tmp &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"#PBS -l walltime="</span>, hpc.walltime, <span class="st">":59:00,pcput="</span>, hpc.walltime, 
    <span class="st">":45:00"</span>, <span class="dt">sep =</span> <span class="st">""</span>)
pbshead &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(pbshead, tmp, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
tmp &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"#PBS -l select="</span>, hpc.select, <span class="st">":ncpus="</span>, hpc.nproc, <span class="st">":mem="</span>, hpc.mem, 
    <span class="dt">sep =</span> <span class="st">""</span>)
pbshead &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(pbshead, tmp, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
pbshead &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(pbshead, <span class="st">"#PBS -j oe"</span>, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
if (!<span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(hpc.array)) pbshead &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(pbshead, <span class="st">"</span><span class="ch">\n</span><span class="st">#PBS -J 1-"</span>, hpc.array, 
    <span class="dt">sep =</span> <span class="st">""</span>)
if (!<span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(hpc.q)) pbshead &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(pbshead, <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"#PBS -q"</span>, hpc.q), <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
pbshead &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(pbshead, hpc.load, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</code></pre></div>
<p>Our header thus looks as follows:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">#!/bin/sh</span>
<span class="co">#PBS -l walltime=71:59:00,pcput=71:45:00</span>
<span class="co">#PBS -l select=1:ncpus=1:mem=6gb</span>
<span class="co">#PBS -j oe</span>
<span class="co">#PBS -J 1-2</span>
<span class="co">#PBS -q pqeelab</span>
<span class="kw">module</span> load intel-suite/2015.1 mpi raxml/8.2.9 mafft/7 anaconda/2.3.0 samtools</code></pre></div>
<p>We are now ready to create an array script, and submit the array job:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># create PBS job array</span>
cmd &lt;-<span class="st"> </span>pty.c[, <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">CASE =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(CASE_ID, <span class="st">")</span><span class="ch">\n</span><span class="st">"</span>, CMD, <span class="st">";;</span><span class="ch">\n</span><span class="st">"</span>)), by =<span class="st"> "CASE_ID"</span>]
cmd &lt;-<span class="st"> </span>cmd[, <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"case $PBS_ARRAY_INDEX in</span><span class="ch">\n</span><span class="st">"</span>, <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(CASE, <span class="dt">collapse =</span> <span class="st">""</span>), 
    <span class="st">"esac"</span>)]
cmd &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(pbshead, cmd, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
<span class="co"># submit job</span>
outfile &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">":"</span>, <span class="st">""</span>, <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"readali"</span>, <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/date.html">date</a></span>(), <span class="dt">split =</span> <span class="st">" "</span>)[[<span class="dv">1</span>]], 
    <span class="dt">collapse =</span> <span class="st">"_"</span>, <span class="dt">sep =</span> <span class="st">""</span>), <span class="st">"sh"</span>, <span class="dt">sep =</span> <span class="st">"."</span>))
outfile &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(pty.args[[<span class="st">"work.dir"</span>]], outfile)
<span class="kw"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(cmd, <span class="dt">file =</span> outfile)
cmd &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"qsub"</span>, outfile)
<span class="kw"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(cmd)
<span class="kw"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.html">system</a></span>(cmd, <span class="dt">intern =</span> <span class="ot">TRUE</span>))</code></pre></div>
</div>
<div id="prepare-bash-scripts-to-reconstruct-deep-sequence-trees" class="section level2">
<h2 class="hasAnchor">
<a href="#prepare-bash-scripts-to-reconstruct-deep-sequence-trees" class="anchor"></a>Prepare bash scripts to reconstruct deep-sequence trees</h2>
<p>As soon as the first read alignments are generated, we can plough ahead and reconstruct viral phylogenies. There is no need to wait until all alignments have been generated. Because we are considering highly overlapping genomic windows across the genome, we will reconstruct one tree per genomic window. The resulting trees across the genome capture not only uncertainty in tree reconstruction, but also differences in viral evolution across the genome, differences in read amplification and mapping, and uncertainty in alignment construction. This is why we prefer to reconstruct trees from overlapping windows, rather than bootstrap trees or samples of posterior tree distributions for a few genomic windows.</p>
<p>We will add the trees to the directory in which the read alignments are, so let us re-set the input directory:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">in.dir &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(HOME, <span class="st">"MRCPopSample_phsc_stage1_output"</span>)</code></pre></div>
<p>Next, we search for read alignments, and check that the corresponding trees have not yet been generated:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># search for read alignments</span>
infiles &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">FI =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(in.dir, <span class="dt">pattern =</span> <span class="st">"fasta$"</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>, 
    <span class="dt">recursive =</span> <span class="ot">TRUE</span>))
infiles[, <span class="st">`</span><span class="dt">:=</span><span class="st">`</span>(FO, <span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"fasta$"</span>, <span class="st">"tree"</span>, FI))]
infiles[, <span class="st">`</span><span class="dt">:=</span><span class="st">`</span>(PTY_RUN, <span class="kw"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"^ptyr([0-9]+)_.*"</span>, <span class="st">"</span><span class="ch">\\</span><span class="st">1"</span>, <span class="kw"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(FI))))]
infiles[, <span class="st">`</span><span class="dt">:=</span><span class="st">`</span>(W_FROM, <span class="kw"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">".*InWindow_([0-9]+)_.*"</span>, <span class="st">"</span><span class="ch">\\</span><span class="st">1"</span>, <span class="kw"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(FI))))]
<span class="co"># check which (if any) trees have already been processed, and remove from</span>
<span class="co"># TODO list</span>
tmp &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">FT =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(out.dir, <span class="dt">pattern =</span> <span class="st">"^ptyr.*tree$"</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>, 
    <span class="dt">recursive =</span> <span class="ot">TRUE</span>))
tmp[, <span class="st">`</span><span class="dt">:=</span><span class="st">`</span>(PTY_RUN, <span class="kw"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"^ptyr([0-9]+)_.*"</span>, <span class="st">"</span><span class="ch">\\</span><span class="st">1"</span>, <span class="kw"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(FT))))]
tmp[, <span class="st">`</span><span class="dt">:=</span><span class="st">`</span>(W_FROM, <span class="kw"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">".*InWindow_([0-9]+)_.*"</span>, <span class="st">"</span><span class="ch">\\</span><span class="st">1"</span>, <span class="kw"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(FT))))]
infiles &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/merge.html">merge</a></span>(infiles, tmp, <span class="dt">by =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"PTY_RUN"</span>, <span class="st">"W_FROM"</span>), <span class="dt">all.x =</span> <span class="dv">1</span>)
infiles &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(infiles, <span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(FT))</code></pre></div>
<p>The read alignments can differ substantially in size, so we expect that the time and memory requirements to generate phylogenies also differ. We will do the easy jobs first, and leave the really big phylogenies that we may want to reconstruct with multiple cpu cores and a lot of RAM until the very end. With this strategy in mind, we order the alignments by the number of reads in each, and associate a ‘CASE_ID’ to each alignment:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tmp &lt;-<span class="st"> </span>infiles[, <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">N_TAXA =</span> <span class="kw"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.html">system</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"grep -c </span><span class="ch">\"</span><span class="st">^&gt;</span><span class="ch">\"</span><span class="st"> "</span>, FI), 
    <span class="dt">intern =</span> <span class="ot">TRUE</span>))), by =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"FI"</span>)]
infiles &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/merge.html">merge</a></span>(infiles, tmp, <span class="dt">by =</span> <span class="st">"FI"</span>)
infiles &lt;-<span class="st"> </span>infiles[<span class="kw"><a href="https://rdrr.io/r/base/order.html">order</a></span>(N_TAXA), ]
infiles[, <span class="st">`</span><span class="dt">:=</span><span class="st">`</span>(CASE_ID, <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(infiles)))]
<span class="kw">setkey</span>(infiles, CASE_ID)</code></pre></div>
Let us take a look at the files to be processed for just one run (in our case, run 100):
<p align="center">
<img src="figures/Stage1.sortedbams.png" alt="Read alignments, sorted by size."></p>
<p>The smallest alignment has 194 reads, which is the alignment of the genomic window that starts at position 4750 of the HXB2 reference genome. The largest alignment has 2377 reads, and corresponds to the genomic window that starts at position 1800 of the HXB2 reference genome.</p>
<p>As before, we will generate UNIX <em>bash</em> scripts to reconstruct trees in parallel, and submit each job to a job scheduler of a high-performance computing system. This time though, we will start with the job specification because the RAxML command specification depends on the job specification. For small read alignments, a single processor will do. Keep in mind that for large alignments, you likely want to change the following job specification.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">infiles &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(infiles, CASE_ID &lt;=<span class="st"> </span><span class="dv">10000</span>)
hpc.load &lt;-<span class="st"> "module load intel-suite/2015.1 mpi raxml/8.2.9"</span>  <span class="co"># make third party requirements available\t </span>
hpc.select &lt;-<span class="st"> </span><span class="dv">1</span>  <span class="co"># number of nodes</span>
hpc.nproc &lt;-<span class="st"> </span><span class="dv">1</span>  <span class="co"># number of processors on node</span>
hpc.walltime &lt;-<span class="st"> </span><span class="dv">71</span>  <span class="co"># walltime</span>
hpc.q &lt;-<span class="st"> </span><span class="ot">NA</span>  <span class="co"># PBS queue</span>
hpc.mem &lt;-<span class="st"> "2gb"</span>  <span class="co"># RAM\t</span>
hpc.array &lt;-<span class="st"> </span>infiles[, <span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(CASE_ID)]  <span class="co"># number of runs for job array\t</span>
<span class="co"># define PBS header for job scheduler. this will depend on your job</span>
<span class="co"># scheduler.</span>
pbshead &lt;-<span class="st"> "#!/bin/sh"</span>
tmp &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"#PBS -l walltime="</span>, hpc.walltime, <span class="st">":59:00,pcput="</span>, hpc.walltime, 
    <span class="st">":45:00"</span>, <span class="dt">sep =</span> <span class="st">""</span>)
pbshead &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(pbshead, tmp, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
tmp &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"#PBS -l select="</span>, hpc.select, <span class="st">":ncpus="</span>, hpc.nproc, <span class="st">":mem="</span>, hpc.mem, 
    <span class="dt">sep =</span> <span class="st">""</span>)
pbshead &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(pbshead, tmp, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
pbshead &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(pbshead, <span class="st">"#PBS -j oe"</span>, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
if (!<span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(hpc.array)) pbshead &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(pbshead, <span class="st">"</span><span class="ch">\n</span><span class="st">#PBS -J 1-"</span>, hpc.array, 
    <span class="dt">sep =</span> <span class="st">""</span>)
if (!<span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(hpc.q)) pbshead &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(pbshead, <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"#PBS -q"</span>, hpc.q), <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
pbshead &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(pbshead, hpc.load, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</code></pre></div>
<p>Depending on the job spec, we define the RAxML command and RAxML arguments:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># raxml.pr &lt;- ifelse(hpc.nproc==1, 'raxmlHPC-AVX','raxmlHPC-PTHREADS-AVX')</span>
<span class="co"># #on newer machines with AVX instructions</span>
raxml.pr &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(hpc.nproc ==<span class="st"> </span><span class="dv">1</span>, <span class="st">"raxmlHPC-SSE3"</span>, <span class="st">"raxmlHPC-PTHREADS-SSE3"</span>)  <span class="co">#on older machines without AVX instructions</span>
raxml.args &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(hpc.nproc ==<span class="st"> </span><span class="dv">1</span>, <span class="st">"-m GTRCAT --HKY85 -p 42 -o REF_B_K03455"</span>, 
    <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"-m GTRCAT --HKY85 -T "</span>, hpc.nproc, <span class="st">" -p 42 -o REF_B_K03455"</span>))</code></pre></div>
<p>Finally, we prepare the UNIX <em>bash</em> scripts to generate a maximum-likelihood deep-sequecen tree for each read alignment:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pty.c &lt;-<span class="st"> </span>infiles[, <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">CMD =</span> <span class="kw"><a href="../reference/raxml.cmd.html">raxml.cmd</a></span>(FI, <span class="dt">outfile =</span> FO, <span class="dt">pr =</span> raxml.pr, <span class="dt">pr.args =</span> raxml.args)), 
    by =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CASE_ID"</span>)]</code></pre></div>
<p>Let us have a look at the commands in one script:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pty.c[<span class="dv">1</span>, <span class="kw"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(CMD)]</code></pre></div>
<p align="center">
<img src="figures/Stage1.RAxMLscript.png" alt="Unix bash script to construct one deep-sequence phylogeny."></p>
<p>In about the middle of the script, we see a call to RAxML, great. As before, this is encapsulated by file management commands to ensure that the script is self-contained, and that the script operates on a local directory in case the data are stored on a central data storage system.</p>
</div>
<div id="run-bash-scripts-to-reconstruct-deep-sequence-trees" class="section level2">
<h2 class="hasAnchor">
<a href="#run-bash-scripts-to-reconstruct-deep-sequence-trees" class="anchor"></a>Run bash scripts to reconstruct deep-sequence trees</h2>
<p>The final task is to submit the <em>bash</em> scripts to a job scheduler of a high-performance computing system:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># make array job</span>
cmd &lt;-<span class="st"> </span>pty.c[, <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">CASE =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(CASE_ID, <span class="st">")</span><span class="ch">\n</span><span class="st">"</span>, CMD, <span class="st">";;</span><span class="ch">\n</span><span class="st">"</span>)), by =<span class="st"> "CASE_ID"</span>]
cmd &lt;-<span class="st"> </span>cmd[, <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"case $PBS_ARRAY_INDEX in</span><span class="ch">\n</span><span class="st">"</span>, <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(CASE, <span class="dt">collapse =</span> <span class="st">""</span>), 
    <span class="st">"esac"</span>)]
cmd &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(pbshead, cmd, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
<span class="co"># submit job</span>
outfile &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">":"</span>, <span class="st">""</span>, <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"trs"</span>, <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/date.html">date</a></span>(), <span class="dt">split =</span> <span class="st">" "</span>)[[<span class="dv">1</span>]], 
    <span class="dt">collapse =</span> <span class="st">"_"</span>, <span class="dt">sep =</span> <span class="st">""</span>), <span class="st">"sh"</span>, <span class="dt">sep =</span> <span class="st">"."</span>))
outfile &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(work.dir, outfile)
<span class="kw"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(cmd, <span class="dt">file =</span> outfile)
cmd &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"qsub"</span>, outfile)
<span class="kw"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(cmd)
<span class="kw"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.html">system</a></span>(cmd, <span class="dt">intern =</span> <span class="ot">TRUE</span>))</code></pre></div>
This time, the job array script is very large with up to 10,000 jobs. It should look similar to the following, which highlights jobs 312 and 313:
<p align="center">
<img src="figures/Stage1.RAxMLArrayscript.png" alt="Unix bash script to construct one deep-sequence phylogeny."></p>
<p>Remember that we generate 10,000 trees for the smallest read alignments with one array job, so we need to repeat the above steps several times. The limit of 10,000 was chosen so that the job array file does not get too large. Also, once you get to really large read alignments, you will likely want to change to run RAxML with multiple cores and more RAM.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#setting-up-the-analysis">Setting up the analysis</a></li>
      <li><a href="#setting-up-phyloscanner-runs-to-test-the-phylogenetic-relationship-of-all-possible-pairs-of-individuals">Setting up <em>phyloscanner</em> runs to test the phylogenetic relationship of all possible pairs of individuals</a></li>
      <li><a href="#prepare-bash-scripts-to-generate-read-alignments">Prepare bash scripts to generate read alignments</a></li>
      <li><a href="#run-bash-scripts-to-generate-read-alignments">Run bash scripts to generate read alignments</a></li>
      <li><a href="#prepare-bash-scripts-to-reconstruct-deep-sequence-trees">Prepare bash scripts to reconstruct deep-sequence trees</a></li>
      <li><a href="#run-bash-scripts-to-reconstruct-deep-sequence-trees">Run bash scripts to reconstruct deep-sequence trees</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Oliver Ratmann.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9000.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
