<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Stage 1 analysis: identify phylogenetically closely related individuals • Phyloscanner.R.utilities</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">Phyloscanner.R.utilities</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Rakai.01.data_description.html">Data description</a>
    </li>
    <li>
      <a href="../articles/Rakai.02.run_phyloscanner.html">Running *phyloscanner* in high-throughput on a population-based sample</a>
    </li>
    <li>
      <a href="../articles/Rakai.03.reconstruct_transmission_networks.html">Reconstructing transmission networks</a>
    </li>
    <li>
      <a href="../articles/Rakai.04.direction_of_transmission.html">Inferring the direction of transmission</a>
    </li>
    <li>
      <a href="../articles/Stage1.create.read.alignments.html">Stage 1 analysis: identify phylogenetically closely related individuals</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/olli0601/Phyloscanner.R.utilities">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Stage 1 analysis: identify phylogenetically closely related individuals</h1>
                        <h4 class="author">Oliver Ratmann</h4>
            
            <h4 class="date">2018-10-30</h4>
          </div>

    
    
<div class="contents">
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>This tutorial describes how to identify phylogenetically closely related individuals with <em>phyloscanner</em> from of a large population-based sample of deep sequences.</p>
</div>
<div id="setting-up-the-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-up-the-analysis" class="anchor"></a>Setting up the analysis</h2>
<p>Let us define input and output directories for our analysis in R.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(Phyloscanner.R.utilities)
HOME &lt;-<span class="st"> "~/sandbox/DeepSeqProjects"</span>
data.dir &lt;-<span class="st"> "~/sandbox/DeepSeqProjects/MRCPopSample_data"</span>
prog.pty &lt;-<span class="st"> "/Users/Oliver/git/phylotypes/phyloscanner_make_trees.py"</span>
in.dir &lt;-<span class="st"> </span><span class="kw">file.path</span>(HOME, <span class="st">"MRCPopSample_phsc_stage1_input"</span>)
work.dir &lt;-<span class="st"> </span><span class="kw">file.path</span>(HOME, <span class="st">"MRCPopSample_phsc_work"</span>)
out.dir &lt;-<span class="st"> </span><span class="kw">file.path</span>(HOME, <span class="st">"MRCPopSample_phsc_stage1_output"</span>)
<span class="kw">dir.create</span>(in.dir, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)
<span class="kw">dir.create</span>(out.dir, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)
<span class="kw">dir.create</span>(work.dir, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>Here, <code>prog.pty</code> is the full path to the <em>phyloscanner</em> program <em>phyloscanner_make_trees.py</em> and <code>work.dir</code> is the name of a temporary directory.</p>
</div>
<div id="setting-up-phyloscanner-runs-to-test-the-phylogenetic-relationship-of-all-possible-pairs-of-individuals" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-up-phyloscanner-runs-to-test-the-phylogenetic-relationship-of-all-possible-pairs-of-individuals" class="anchor"></a>Setting up <em>phyloscanner</em> runs to test the phylogenetic relationship of all possible pairs of individuals</h2>
<p>Next, we define a large number of <em>phyloscanner</em> runs to test the phylogenetic relationship of all possible pairs of individuals. We start by generating a list of all bam files in our data directory (<code>data.dir</code>), and then we identify the corresponding individuals with a regular expression:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dbam &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">BAM =</span> <span class="kw">list.files</span>(data.dir, <span class="dt">pattern =</span> <span class="st">".*remap</span><span class="ch">\\</span><span class="st">.bam$"</span>, <span class="dt">full.names =</span> <span class="ot">FALSE</span>, 
    <span class="dt">recursive =</span> <span class="ot">TRUE</span>))
regex.person &lt;-<span class="st"> "^([A-Z0-9]+-[A-Z0-9]+)-.*$"</span>
dbam[, <span class="st">`</span><span class="dt">:=</span><span class="st">`</span>(IND, <span class="kw">gsub</span>(regex.person, <span class="st">"</span><span class="ch">\\</span><span class="st">1"</span>, BAM))]</code></pre></div>
The data.table <code>dbam</code> should look similar to the following:
<p align="center">
<img src="figures/Stage1.bams.png" alt="List of bam files, stage 1 of analysis of large population-based sample."></p>
<p>We then group individuals for phyloscanner analyses, so that phylogenetic linkage between every pair of individuals is assessed at least once. Specifically, individuals are grouped into batches of specified size, and then, all possible pairs of batches are formed. Each of these pairs of batches defines a group of individuals between whom phylogenetic linkages are assessed in one phyloscanner run. Thus, the number of individuals in each group is twice the batch size:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tmp &lt;-<span class="st"> </span><span class="kw">unique</span>(dbam$IND)
pty.runs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/phsc.define.stage1.analyses.html">phsc.define.stage1.analyses</a></span>(tmp, <span class="dt">batch.size =</span> <span class="dv">50</span>)
<span class="co"># add BAM files per individual</span>
pty.runs &lt;-<span class="st"> </span><span class="kw">merge</span>(pty.runs, <span class="kw">subset</span>(dbam, <span class="dt">select =</span> <span class="kw">c</span>(IND, BAM)), <span class="dt">by =</span> <span class="st">"IND"</span>, 
    <span class="dt">allow.cartesian =</span> <span class="ot">TRUE</span>)
<span class="kw">setkey</span>(pty.runs, PTY_RUN)
<span class="co"># save to file</span>
outfile &lt;-<span class="st"> "phsc_runs_MRC_stage1_n2531_181026.csv"</span>
<span class="kw">write.csv</span>(pty.runs, <span class="dt">file =</span> <span class="kw">file.path</span>(in.dir, outfile), <span class="dt">row.names =</span> <span class="ot">FALSE</span>)</code></pre></div>
The data.table <code>pty.runs</code> should look similar to the following:
<p align="center">
<img src="figures/Stage1.predefinedruns.png" alt="List of phyloscanner runs, stage 1 of analysis of large population-based sample."></p>
<p>The data.table specifies a large number of <em>phyloscanner</em> runs. All individuals (‘IND’) are grouped in batches (‘BATCH’), and two batches are paired to evaluate the phylogenetic relationship of the associated individuals in a particular analysis (‘PTY_RUN’).</p>
</div>
<div id="prepare-bash-scripts-to-generate-read-alignments" class="section level2">
<h2 class="hasAnchor">
<a href="#prepare-bash-scripts-to-generate-read-alignments" class="anchor"></a>Prepare bash scripts to generate read alignments</h2>
<p>The next step is to define the input arguments generate read alignments with <em>phyloscanner</em>. Please see the <em>phyloscanner</em> manual for details. The default arguments that were used for analysis of the Rakai population-based sample are as follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pty.select &lt;-<span class="st"> </span><span class="dv">1</span>:<span class="dv">2</span>  <span class="co">#for demo purposes, we select the first two runs. Set to NA.</span>
pty.args &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">prog.pty =</span> prog.pty, <span class="dt">prog.mafft =</span> <span class="st">"mafft"</span>, <span class="dt">data.dir =</span> data.dir, 
    <span class="dt">work.dir =</span> work.dir, <span class="dt">out.dir =</span> out.dir, <span class="dt">alignments.file =</span> <span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">"Phyloscanner.R.utilities"</span>, 
        <span class="st">"HIV1_compendium_AD_B_CPX_v2.fasta"</span>), <span class="dt">alignments.root =</span> <span class="st">"REF_CPX_AF460972"</span>, 
    <span class="dt">alignments.pairwise.to =</span> <span class="st">"REF_B_K03455"</span>, <span class="dt">window.automatic =</span> <span class="st">""</span>, <span class="dt">merge.threshold =</span> <span class="dv">2</span>, 
    <span class="dt">min.read.count =</span> <span class="dv">1</span>, <span class="dt">quality.trim.ends =</span> <span class="dv">23</span>, <span class="dt">min.internal.quality =</span> <span class="dv">23</span>, <span class="dt">merge.paired.reads =</span> <span class="ot">TRUE</span>, 
    <span class="dt">no.trees =</span> <span class="ot">TRUE</span>, <span class="dt">dont.check.duplicates =</span> <span class="ot">FALSE</span>, <span class="dt">dont.check.recombination =</span> <span class="ot">TRUE</span>, 
    <span class="dt">win =</span> <span class="kw">c</span>(<span class="dv">800</span>, <span class="dv">9400</span>, <span class="dv">25</span>, <span class="dv">250</span>), <span class="dt">keep.overhangs =</span> <span class="ot">FALSE</span>, <span class="dt">mem.save =</span> <span class="dv">0</span>, <span class="dt">verbose =</span> <span class="ot">TRUE</span>, 
    <span class="dt">select =</span> pty.select  <span class="co">#of 240</span>
)
<span class="kw">save</span>(pty.args, <span class="dt">file =</span> <span class="kw">file.path</span>(in.dir, <span class="st">"phsc_args_stage1_create_read_alignments.rda"</span>))</code></pre></div>
<p>Next, we will prepare UNIX bash scripts to process the data in parallel. Each bash script will consist of UNIX commands to generate alignments of deep-sequence viral reads for one phyloscanner run (‘PTY_RUN’). All scripts, and hence all runs can be performed in parallel. Each alignment contains viral reads of the individuals in one run, and all reads overlap a particular genomic window of the HIV-1 genome:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># check which (if any) batches have already been processed, and remove from</span>
<span class="co"># TODO list</span>
tmp &lt;-<span class="st"> </span><span class="kw">data.table</span>(<span class="dt">FILE_FASTA =</span> <span class="kw">list.files</span>(out.dir, <span class="dt">pattern =</span> <span class="st">"^ptyr[0-9]+_"</span>, 
    <span class="dt">full.names =</span> <span class="ot">TRUE</span>))
tmp[, <span class="st">`</span><span class="dt">:=</span><span class="st">`</span>(PTY_RUN, <span class="kw">as.integer</span>(<span class="kw">gsub</span>(<span class="st">"ptyr([0-9]+)_.*"</span>, <span class="st">"</span><span class="ch">\\</span><span class="st">1"</span>, <span class="kw">basename</span>(FILE_FASTA))))]
pty.runs &lt;-<span class="st"> </span><span class="kw">merge</span>(pty.runs, tmp, <span class="dt">by =</span> <span class="st">"PTY_RUN"</span>, <span class="dt">all.x =</span> <span class="dv">1</span>)
pty.runs &lt;-<span class="st"> </span><span class="kw">subset</span>(pty.runs, <span class="kw">is.na</span>(FILE_FASTA))
<span class="co"># search for bam files and references and merge with runs</span>
<span class="kw">setkey</span>(pty.runs, PTY_RUN)
pty.runs &lt;-<span class="st"> </span><span class="kw">subset</span>(pty.runs, <span class="dt">select =</span> <span class="kw">c</span>(PTY_RUN, IND))
tmp &lt;-<span class="st"> </span><span class="kw">phsc.find.bam.and.references</span>(pty.args[[<span class="st">"data.dir"</span>]], <span class="dt">regex.person =</span> <span class="st">"^([A-Z0-9]+-[A-Z0-9]+)-.*$"</span>)
pty.runs &lt;-<span class="st"> </span><span class="kw">merge</span>(pty.runs, tmp, <span class="dt">by =</span> <span class="st">"IND"</span>)
<span class="co"># create UNIX bash scripts</span>
<span class="kw">setnames</span>(pty.runs, <span class="kw">c</span>(<span class="st">"IND"</span>, <span class="st">"SAMPLE"</span>), <span class="kw">c</span>(<span class="st">"UNIT_ID"</span>, <span class="st">"SAMPLE_ID"</span>))
pty.c &lt;-<span class="st"> </span><span class="kw"><a href="../reference/phsc.cmd.phyloscanner.multi.html">phsc.cmd.phyloscanner.multi</a></span>(pty.runs, pty.args)
pty.c[, <span class="st">`</span><span class="dt">:=</span><span class="st">`</span>(CASE_ID, <span class="kw">seq_len</span>(<span class="kw">nrow</span>(pty.c)))]</code></pre></div>
<p>Let us take a look at the first batch script:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pty.c[<span class="dv">1</span>, <span class="kw">cat</span>(CMD)]</code></pre></div>
<p align="center">
<img src="figures/Stage1.createbamscmd.png" alt="Bash script to generate read alignments, stage 1 of analysis of large population-based sample."></p>
<p>We see the familiar call to <code>phyloscanner_make_trees.py</code>, and we suppressed tree reconstruction with the input command <code>--no-trees</code>. There is also a bit of file copying before/after. This ensures that the script is self-contained, and that it operates on a local directory in case the data are stored on a central data storage system.</p>
</div>
<div id="run-bash-scripts-to-generate-read-alignments" class="section level2">
<h2 class="hasAnchor">
<a href="#run-bash-scripts-to-generate-read-alignments" class="anchor"></a>Run bash scripts to generate read alignments</h2>
<p>The next task is submit the bash scripts to a job scheduling system on a high performance computing environment. We first define a PBS header for the job scheduling system, add the header to each script, and then submit each script to the job scheduling system. The exact form of the PBS header depends on your job scheduler, below is an example that works at Imperial.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hpc.load &lt;-<span class="st"> "module load R/3.3.3"</span>  <span class="co"># make R available </span>
hpc.select &lt;-<span class="st"> </span><span class="dv">1</span>  <span class="co"># number of nodes</span>
hpc.nproc &lt;-<span class="st"> </span><span class="dv">1</span>  <span class="co"># number of processors on node</span>
hpc.walltime &lt;-<span class="st"> </span><span class="dv">15</span>  <span class="co"># walltime</span>
hpc.q &lt;-<span class="st"> "pqeelab"</span>  <span class="co"># PBS queue</span>
hpc.mem &lt;-<span class="st"> "6gb"</span>  <span class="co"># RAM</span>
hpc.array &lt;-<span class="st"> </span>pty.c[, <span class="kw">max</span>(CASE_ID)]  <span class="co"># number of runs for job array\t</span>
<span class="co"># define PBS header for job scheduler. this will depend on your job</span>
<span class="co"># scheduler.</span>
pbshead &lt;-<span class="st"> "#!/bin/sh"</span>
tmp &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">"#PBS -l walltime="</span>, hpc.walltime, <span class="st">":59:00,pcput="</span>, hpc.walltime, 
    <span class="st">":45:00"</span>, <span class="dt">sep =</span> <span class="st">""</span>)
pbshead &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, tmp, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
tmp &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">"#PBS -l select="</span>, hpc.select, <span class="st">":ncpus="</span>, hpc.nproc, <span class="st">":mem="</span>, hpc.mem, 
    <span class="dt">sep =</span> <span class="st">""</span>)
pbshead &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, tmp, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
pbshead &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, <span class="st">"#PBS -j oe"</span>, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
if (!<span class="kw">is.na</span>(hpc.array)) pbshead &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, <span class="st">"</span><span class="ch">\n</span><span class="st">#PBS -J 1-"</span>, hpc.array, 
    <span class="dt">sep =</span> <span class="st">""</span>)
if (!<span class="kw">is.na</span>(hpc.q)) pbshead &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, <span class="kw">paste</span>(<span class="st">"#PBS -q"</span>, hpc.q), <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
pbshead &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, hpc.load, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</code></pre></div>
<p>Our header thus looks as follows:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">#!/bin/sh</span>
<span class="co">#PBS -l walltime=71:59:00,pcput=71:45:00</span>
<span class="co">#PBS -l select=1:ncpus=1:mem=6gb</span>
<span class="co">#PBS -j oe</span>
<span class="co">#PBS -J 1-2</span>
<span class="co">#PBS -q pqeelab</span>
<span class="kw">module</span> load intel-suite/2015.1 mpi raxml/8.2.9 mafft/7 anaconda/2.3.0 samtools</code></pre></div>
<p>We are now ready to create an array script, and submit the array job:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># create PBS job array</span>
cmd &lt;-<span class="st"> </span>pty.c[, <span class="kw">list</span>(<span class="dt">CASE =</span> <span class="kw">paste0</span>(CASE_ID, <span class="st">")</span><span class="ch">\n</span><span class="st">"</span>, CMD, <span class="st">";;</span><span class="ch">\n</span><span class="st">"</span>)), by =<span class="st"> "CASE_ID"</span>]
cmd &lt;-<span class="st"> </span>cmd[, <span class="kw">paste0</span>(<span class="st">"case $PBS_ARRAY_INDEX in</span><span class="ch">\n</span><span class="st">"</span>, <span class="kw">paste0</span>(CASE, <span class="dt">collapse =</span> <span class="st">""</span>), 
    <span class="st">"esac"</span>)]
cmd &lt;-<span class="st"> </span><span class="kw">paste</span>(pbshead, cmd, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
<span class="co"># submit job</span>
outfile &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">":"</span>, <span class="st">""</span>, <span class="kw">paste</span>(<span class="st">"readali"</span>, <span class="kw">paste</span>(<span class="kw">strsplit</span>(<span class="kw">date</span>(), <span class="dt">split =</span> <span class="st">" "</span>)[[<span class="dv">1</span>]], 
    <span class="dt">collapse =</span> <span class="st">"_"</span>, <span class="dt">sep =</span> <span class="st">""</span>), <span class="st">"sh"</span>, <span class="dt">sep =</span> <span class="st">"."</span>))
outfile &lt;-<span class="st"> </span><span class="kw">file.path</span>(pty.args[[<span class="st">"work.dir"</span>]], outfile)
<span class="kw">cat</span>(cmd, <span class="dt">file =</span> outfile)
cmd &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">"qsub"</span>, outfile)
<span class="kw">cat</span>(cmd)
<span class="kw">cat</span>(<span class="kw">system</span>(cmd, <span class="dt">intern =</span> <span class="ot">TRUE</span>))</code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#setting-up-the-analysis">Setting up the analysis</a></li>
      <li><a href="#setting-up-phyloscanner-runs-to-test-the-phylogenetic-relationship-of-all-possible-pairs-of-individuals">Setting up <em>phyloscanner</em> runs to test the phylogenetic relationship of all possible pairs of individuals</a></li>
      <li><a href="#prepare-bash-scripts-to-generate-read-alignments">Prepare bash scripts to generate read alignments</a></li>
      <li><a href="#run-bash-scripts-to-generate-read-alignments">Run bash scripts to generate read alignments</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Oliver Ratmann.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
