<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Generate bash command for a single phyloscanner run — phsc.cmd.phyloscanner.one • Phyloscanner.R.utilities</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">Phyloscanner.R.utilities</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Rakai.01.data_description.html">Data description</a>
    </li>
    <li>
      <a href="../articles/Rakai.02.run_phyloscanner.html">Running *phyloscanner* in high-throughput on a population-based sample</a>
    </li>
    <li>
      <a href="../articles/Rakai.03.reconstruct_transmission_networks.html">Reconstructing transmission networks</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/olli0601/Phyloscanner.R.utilities">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate bash command for a single phyloscanner run</h1>
    </div>

    
    <p>This function generates bash commands for a single phyloscanner run, that can be called via 'system' in R, or written to file to run on a UNIX system.</p>
    

    <pre class="usage"><span class='fu'>phsc.cmd.phyloscanner.one</span>(<span class='no'>pty.args</span>, <span class='no'>file.input</span>, <span class='no'>file.patient</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>pty.args</th>
      <td><p>List of phyloscanner input variables. See examples.</p></td>
    </tr>
    <tr>
      <th>file.input</th>
      <td><p>File name of the file that contains the list of bam files, reference files, and potentially aliases</p></td>
    </tr>
    <tr>
      <th>file.patient</th>
      <td><p>File name of the file that contains the list of unique individuals/units to which the bam files correspond. Multiple bam files are allowed per individual/unit.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Character string of phyloscanner commands.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>#	setup input, output and working directories</span>
<span class='no'>pty.data.dir</span>            <span class='kw'>&lt;-</span> <span class='st'>'/Users/Oliver/duke/2016_PANGEAphylotypes/data'</span>
<span class='no'>work.dir</span>                        <span class='kw'>&lt;-</span> <span class='fu'>getwd</span>()
<span class='no'>out.dir</span>                         <span class='kw'>&lt;-</span> <span class='fu'>getwd</span>()
<span class='co'>#	link to phyloscanner.py</span>
<span class='no'>prog.pty</span>                        <span class='kw'>&lt;-</span> <span class='st'>'/work/or105/libs/phylotypes/phyloscanner.py'</span>
<span class='co'>#   input file with reference to bams and refs</span>
<span class='no'>file.bam</span>            <span class='kw'>&lt;-</span> <span class='st'>'XXX'</span>
<span class='no'>file.ref</span>            <span class='kw'>&lt;-</span> <span class='st'>'YYY'</span>
<span class='co'>#	define phyloscanner arguments</span>
<span class='no'>pty.args</span>                        <span class='kw'>&lt;-</span> <span class='fu'>list</span>(        <span class='kw'>prog.pty</span><span class='kw'>=</span><span class='no'>prog.pty</span>,
                                <span class='kw'>prog.mafft</span><span class='kw'>=</span><span class='st'>'mafft'</span>,
                                <span class='kw'>prog.raxml</span><span class='kw'>=</span><span class='st'>'"raxmlHPC-AVX -m GTRCAT"'</span>,
                                <span class='kw'>data.dir</span><span class='kw'>=</span><span class='no'>pty.data.dir</span>,
                                <span class='kw'>work.dir</span><span class='kw'>=</span><span class='no'>work.dir</span>,
                                <span class='kw'>out.dir</span><span class='kw'>=</span><span class='no'>out.dir</span>,
                                <span class='kw'>alignments.file</span><span class='kw'>=</span><span class='fu'>system.file</span>(<span class='kw'>package</span><span class='kw'>=</span><span class='st'>"phyloscan"</span>, <span class='st'>"HIV1_compendium_AD_B_CPX_v2.fasta"</span>),
                                <span class='kw'>alignments.root</span><span class='kw'>=</span><span class='st'>'REF_CPX_AF460972'</span>,
                                <span class='kw'>alignments.pairwise.to</span><span class='kw'>=</span><span class='st'>'REF_B_K03455'</span>,
                                <span class='kw'>window.automatic</span><span class='kw'>=</span> <span class='st'>''</span>,
                                <span class='kw'>merge.threshold</span><span class='kw'>=</span><span class='fl'>0</span>,
                                <span class='kw'>min.read.count</span><span class='kw'>=</span><span class='fl'>1</span>,
                                <span class='kw'>quality.trim.ends</span><span class='kw'>=</span><span class='fl'>23</span>,
                                <span class='kw'>min.internal.quality</span><span class='kw'>=</span><span class='fl'>23</span>,
                                <span class='kw'>merge.paired.reads</span><span class='kw'>=</span><span class='fl'>TRUE</span>,
                                <span class='kw'>no.trees</span><span class='kw'>=</span><span class='fl'>FALSE</span>,
                                <span class='kw'>dont.check.duplicates</span><span class='kw'>=</span><span class='fl'>FALSE</span>,
                                <span class='kw'>num.bootstraps</span><span class='kw'>=</span><span class='fl'>1</span>,
                                <span class='kw'>all.bootstrap.trees</span><span class='kw'>=</span><span class='fl'>TRUE</span>,
                                <span class='kw'>min.ureads.individual</span><span class='kw'>=</span><span class='fl'>NA</span>,
                                <span class='kw'>win</span><span class='kw'>=</span><span class='fu'>c</span>(<span class='fl'>800</span>,<span class='fl'>2400</span>,<span class='fl'>250</span>,<span class='fl'>250</span>),
                                <span class='kw'>keep.overhangs</span><span class='kw'>=</span><span class='fl'>FALSE</span>,
                                <span class='kw'>duplicated.raw.threshold</span><span class='kw'>=</span><span class='fl'>3</span>,
                                <span class='kw'>duplicated.ratio.threshold</span><span class='kw'>=</span><span class='fl'>1</span>/<span class='fl'>200</span>,
                                <span class='kw'>select</span><span class='kw'>=</span><span class='fl'>NA</span>)
<span class='co'>#	generate bash script</span>
<span class='no'>cmd</span>                 <span class='kw'>&lt;-</span> <span class='fu'>phsc.cmd.phyloscanner.one</span>(<span class='no'>file.bam</span>, <span class='no'>file.ref</span>, <span class='no'>pty.args</span>)</div><div class='output co'>#&gt; <span class='error'>Error: is.character(file.patient) is not TRUE</span></div><div class='input'><span class='co'>#	print bash script to screen</span>
<span class='fu'>cat</span>(<span class='no'>cmd</span>)</div><div class='output co'>#&gt; <span class='error'>Error in cat(cmd): object 'cmd' not found</span></div><div class='input'>#   this can be started eg through system on a UNIX machine, or written to file and then started manually (recommended)
#system(cmd)
#file.out <- 'ZZZ'
#cat(cmd, file=file.out)
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Oliver Ratmann.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
